<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>get_bags</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">get_bags</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this vignette we will explore the function
<code>get_bags()</code>. This function is used internally in both
<code>aurora_pheno()</code> and <code>aurora_GWAS()</code> but can be
also accessed upon installing <em>aurora</em>. The function takes a
dataset of microbial strains and returns a bootstrapped dataset that is
adjusted for the population structure. <code>get_bags()</code> has many
possible uses and we will demonstrate a few here. ## Description of
function <code>get_bags()</code> This function requires two inputs. The
phenotype matrix (<code>pheno_mat</code>) which has to be in the same
format as the one used in <code>aurora_pheno()</code> and
<code>aurora_GWAS()</code> and a phylogenetic reconstruction. This can
be either a phylogenetic tree loaded as a phylo object or a distance
matrix that contains phylogenetic pairwise distances (core genome amino
acid identities (cAAI) or average nucleotide identities (ANI) converted
to distances). ## Objective Example 1: <code>aurora_pheno()</code> uses
a grid search to fine-tune hyperparameters for Random Forest and
AdaBoost. However, this process is very time consuming. Additionally,
the user is only restricted to the parameters that are in the grid that
<code>aurora_pheno()</code> uses (see supplementary materials in the
aurora paper). However, you can use <code>get_bags()</code> to get the
adjusted datasets and then use your favorite hyperparameter search
method to get the best results.</p>
<p>Example 2: <code>aurora_pheno()</code> uses four machine learning
algorithms to identify the mislabeled strains. However, these machine
learning algorithms may not be optimal for your dataset and you may want
to explore other options. Machine learning algorithms often learn the
phylogenetic background rather than the existing causal features. To
minimize this issue we will first use <code>get_bags()</code> to obtain
multiple bootstrapped datasets that are adjusted for phylogenetic
non-independence and then construct corresponding XGBoost models. The
results from these models will be combined to obtain the most probable
causal features and to identify potentially mislabeled strains (a simple
implementation of <code>aurora_pheno()</code> and
<code>aurora_GWAS()</code>).</p>
<p>Example 3: In scientific literature I often encounter plots that are
showing cluster analysis of microbial strains. The results of the
analysis are then displayed in two dimensions. Such plots are supposed
to convince us that <em>i.e.,</em> cluster A is more associated with
trait B. Moreover, some dimensionality reduction techniques, like PCA,
aims to capture the maximum variability in the dataset. However,
population structure and sampling bias can distort such plot. We will
use function <code>get_bags()</code> to reduce the influence of
population structure and sampling bias. First we will obtain a
population adjusted dataset and then we will calculate Jaccard distance
between all strains. The resulting distance matrix will be plotted to
two dimensions.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(aurora)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(mlr)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(randomForest)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(xgboost)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(proxy)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">library</span>(FactoMineR)</span></code></pre></div>
<p>Now let’s load some data. We will use <em>Limosilactobacillus
reuteri</em> datset. This dataset contains 207 strains from five
different hosts. As a feature table we will use pangenome matrix
produced by <a href="https://github.com/gtonkinhill/panaroo">Panaroo</a>. We will first
need to convert this pangenome into a binary format. Additionally, we
will filter out the most common (present in more than 95% of all
strains) and the rarest (present in less than 10% of all strains)
genes.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;pheno_mat_reuteri&quot;</span>) <span class="co"># loads the phenotype for each strain</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;bin_mat_reuteri&quot;</span>) <span class="co"># loads the pangenome matrix</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;tree_reuteri&quot;</span>) <span class="co"># loads the core-genome phylogenetic tree</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># convert the pangenome matrix to binary matrix</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(bin_mat)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>fut_row_names <span class="ot">&lt;-</span> bin_mat<span class="sc">$</span>Gene</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> bin_mat[, <span class="sc">-</span><span class="dv">1</span><span class="sc">:-</span><span class="dv">14</span>]</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> bin_mat <span class="sc">%&gt;%</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="fu">mutate_all</span>(<span class="sc">~</span><span class="fu">replace</span>(., <span class="fu">is.na</span>(.), <span class="dv">0</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  <span class="fu">mutate_all</span>(<span class="sc">~</span><span class="fu">replace</span>(., .<span class="sc">!=</span><span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> <span class="fu">sapply</span>(bin_mat, as.numeric)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(bin_mat)</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="fu">colnames</span>(bin_mat) <span class="ot">&lt;-</span> fut_row_names</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> bin_mat[<span class="fu">match</span>(pheno_mat<span class="sc">$</span>ids, <span class="fu">rownames</span>(bin_mat)), ]</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>no_gf <span class="ot">&lt;-</span> <span class="fu">nrow</span>(bin_mat)</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>cutoff <span class="ot">&lt;-</span> no_gf <span class="sc">*</span> (<span class="dv">95</span><span class="sc">/</span><span class="dv">100</span>) <span class="co"># remove abundat features</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>select_ID_up <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colSums</span>(bin_mat) <span class="sc">&lt;</span> cutoff)</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>cutoff <span class="ot">&lt;-</span> no_gf <span class="sc">*</span> (<span class="dv">5</span><span class="sc">/</span><span class="dv">100</span>) <span class="co"># remove rare features</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a>non_select_ID_low <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colSums</span>(bin_mat) <span class="sc">&gt;</span> cutoff)</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>bin_mat <span class="ot">&lt;-</span> bin_mat[, <span class="fu">intersect</span>(select_ID_up, non_select_ID_low)]</span></code></pre></div>
</div>
<div id="example-1" class="section level1">
<h1>Example 1</h1>
<p>With the data loaded we can run <code>get_bags()</code> to obtain the
bootstrapped dataset. <em>aurora</em> uses two algorithms to bootstrap
the data: <code>phylogenetic_walk</code> and <code>random_walk</code>.
This is controlled by a parameter <code>bagging</code>. If you wish to
observe the distinctions between the two algorithms, please refer to the
supplementary materials of the <em>aurora</em> paper. Both algorithms
are stochastic and therefore you should either repeat the hyperparameter
tuning multiple times (each time with different bootstrapped dataset) or
you can simply set the parameter <code>bag_size</code> to a large
number. This parameter controls how many strains from each class will be
selected to the bootstrapped dataset. If you want to repeat the
parameter search multiple times, you can get multiple bags with just one
run of <code>get_bags()</code> by changing the parameter
<code>no_rf</code>. In this case, we will set <code>bag_size</code> to a
large number and then use random hyperparameter search to find the best
hyperparameters for random forest model.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Get the train dataset</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>bag <span class="ot">&lt;-</span> <span class="fu">get_bags</span>(pheno_mat,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>                <span class="at">dist_mat_or_tree =</span> tree, </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                <span class="at">bag_size =</span> <span class="fu">c</span>(<span class="dv">200</span>, <span class="dv">200</span>, <span class="dv">200</span>, <span class="dv">200</span>, <span class="dv">200</span>) <span class="co"># number of strains in each class</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                )</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># The number of strains in each class should be balanced for Random Forest</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Get bags also outputs $count which shows the number of strain repetitions</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># Construct new pheno_mat and bin_mat</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>boot_pheno_mat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ids =</span> bag<span class="sc">$</span>bags[[<span class="dv">1</span>]],</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                             <span class="at">pheno =</span> pheno_mat<span class="sc">$</span>pheno[<span class="fu">match</span>(bag<span class="sc">$</span>bags[[<span class="dv">1</span>]], pheno_mat<span class="sc">$</span>ids)])</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>boot_bin_mat<span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(bin_mat[<span class="fu">match</span>(bag<span class="sc">$</span>bags[[<span class="dv">1</span>]], <span class="fu">rownames</span>(bin_mat)), ])</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>boot_bin_mat<span class="sc">$</span>host <span class="ot">&lt;-</span> boot_pheno_mat<span class="sc">$</span>pheno</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co"># gene names from panaroo are not valid column names</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>gene_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(boot_bin_mat)[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(boot_bin_mat)<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="fu">colnames</span>(boot_bin_mat)[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(boot_bin_mat)<span class="sc">-</span><span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(<span class="st">&quot;gene&quot;</span>, <span class="fu">ncol</span>(boot_bin_mat)<span class="sc">-</span><span class="dv">1</span>), <span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(boot_bin_mat)<span class="sc">-</span><span class="dv">1</span>))</span></code></pre></div>
<p>Now we are ready to run the hyperparameter tuning. Any method can be
used (random search, manual search, gird search, Bayesian optimization,
etc.). In this case, random search will be used. This method is
implemented in package <code>mlr</code>. We need to choose a measurement
of the models’ performance. In real scenario, you might want to use <a href="https://towardsdatascience.com/understanding-auc-roc-curve-68b2303cc9c5">AUC</a>
or an <a href="https://towardsdatascience.com/the-f1-score-bec2bbc38aa6">F1
Score</a> however let’s keep things simple here and use accuracy. Keep
in mind that the grid search that <code>aurora_pheno()</code> uses takes
a long time to complete (especially if the your dataset has over 500
strains, over 10,000 features and more than two classes). In such a case
you are forced to find your own hyperparameters.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>task <span class="ot">&lt;-</span> <span class="fu">makeClassifTask</span>(<span class="at">data =</span> boot_bin_mat, <span class="at">target =</span> <span class="st">&quot;host&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># define hyperparameter search space</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>param_set <span class="ot">&lt;-</span> <span class="fu">makeParamSet</span>(</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="fu">makeIntegerParam</span>(<span class="st">&quot;sampsize&quot;</span>, <span class="at">lower =</span> <span class="dv">50</span>, <span class="at">upper =</span> <span class="dv">200</span>),</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="fu">makeIntegerParam</span>(<span class="st">&quot;ntree&quot;</span>, <span class="at">lower =</span> <span class="dv">50</span>, <span class="at">upper =</span> <span class="dv">500</span>),</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="fu">makeIntegerParam</span>(<span class="st">&quot;mtry&quot;</span>, <span class="at">lower =</span> <span class="dv">10</span>, <span class="at">upper =</span> <span class="dv">1000</span>), </span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="fu">makeIntegerParam</span>(<span class="st">&quot;maxnodes&quot;</span>, <span class="at">lower =</span> <span class="dv">4</span>, <span class="at">upper =</span> <span class="dv">20</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  )</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">makeTuneControlRandom</span>(<span class="at">maxit =</span> <span class="dv">10</span>) <span class="co"># set search method as random</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">tuneParams</span>(</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>  <span class="fu">makeLearner</span>(<span class="st">&quot;classif.randomForest&quot;</span>),</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>  task,</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>  <span class="at">resampling =</span> <span class="fu">makeResampleDesc</span>(<span class="st">&quot;CV&quot;</span>, <span class="at">iters =</span> <span class="dv">4</span>),  <span class="co"># 4-fold cross-validation</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="at">measures =</span> <span class="fu">list</span>(acc),  <span class="co"># Use accuracy as the performance measure</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="at">par.set =</span> param_set,</span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>  <span class="at">control =</span> ctrl</span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>)</span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>best_params <span class="ot">&lt;-</span> result<span class="sc">$</span>x</span></code></pre></div>
<p>The best hyperparameters are now in list <code>best_params</code>and
we can use it to run <code>aurora_pheno()</code> (run only with random
forest).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>results_aurora <span class="ot">&lt;-</span> <span class="fu">aurora_pheno</span>(<span class="at">pheno_mat =</span> <span class="fu">data</span>(<span class="st">&quot;pheno_mat_reuteri&quot;</span>),</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>                               <span class="at">bin_mat =</span> <span class="fu">data</span>(<span class="st">&quot;bin_mat_reuteri&quot;</span>),</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                               <span class="at">tree =</span> <span class="fu">data</span>(<span class="st">&quot;tree_reuteri&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                               <span class="at">fit_parameters =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>                               <span class="at">sampsize =</span> best_params<span class="sc">$</span>sampsize,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>                               <span class="at">mtry =</span> best_params<span class="sc">$</span>mtry,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>                               <span class="at">ntree =</span> best_params<span class="sc">$</span>ntree,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>                               <span class="at">maxnodes =</span> best_params<span class="sc">$</span>maxnodes,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>                               <span class="at">ovr_log_reg =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>                               <span class="at">adaboost =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>                               <span class="at">CART =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
</div>
<div id="example-2" class="section level1">
<h1>Example 2:</h1>
<p>In this example we will use the function <code>get_bags()</code> to
get 10 bootstrapped datasets. This datasets will be used as training
datasets to construct 10 corresponding XGBoost models. We don’t want the
training dataset to be too large. This could lead to extensive strain
repetition which could result in XGBoost models that learned the
features present in those repeating strains.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Get the train dataset</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>bag <span class="ot">&lt;-</span> <span class="fu">get_bags</span>(pheno_mat,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                <span class="at">dist_mat_or_tree =</span> tree, </span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                <span class="at">bag_size =</span> <span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">20</span>, <span class="dv">20</span>, <span class="dv">20</span>, <span class="dv">20</span>), <span class="co"># number of strains in each class</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>                <span class="at">no_rf =</span> <span class="dv">10</span>, <span class="co"># specify the number of datasets</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>                <span class="at">bagging =</span> <span class="st">&quot;phylogenetic_walk&quot;</span> <span class="co"># specify the bagging algorithm</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>                )</span></code></pre></div>
<p>Now that the 10 training datasets were obtained we can start building
the XGBoost models. Once an XGBoost model is built we can pull out the
importances (feature gains in this case) and we can also use the model
to predict classification probabilities of the original dataset. In case
of a multiclass problem the feature importances will not make much
sense. Thus, let’s reduce the number of classes to two: rodent and the
rest.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># make a list that will hold all the feature importances</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>importances <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="dv">10</span>)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co"># make a list that will hold all the classification probabilities</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>probs <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">&quot;list&quot;</span>, <span class="dv">10</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co"># gene names from panaroo are not valid column names</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>gene_names <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bin_mat)[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(bin_mat)]</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>bin_mat_adjust <span class="ot">&lt;-</span> bin_mat</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="fu">colnames</span>(bin_mat_adjust)[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(bin_mat_adjust)] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(<span class="st">&quot;gene&quot;</span>, <span class="fu">ncol</span>(bin_mat_adjust)), <span class="dv">1</span><span class="sc">:</span>(<span class="fu">ncol</span>(bin_mat_adjust)))</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="fu">names</span>(gene_names) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bin_mat_adjust)</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>) {</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>  <span class="co"># Construct new pheno_mat and bin_mat</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  boot_pheno_mat <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ids =</span> bag<span class="sc">$</span>bags[[i]],</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>                             <span class="at">pheno =</span> pheno_mat<span class="sc">$</span>pheno[<span class="fu">match</span>(bag<span class="sc">$</span>bags[[i]], pheno_mat<span class="sc">$</span>ids)])</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>  boot_pheno_mat<span class="sc">$</span>pheno <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(boot_pheno_mat<span class="sc">$</span>pheno <span class="sc">==</span> <span class="st">&quot;rodent&quot;</span>, <span class="dv">1</span>, <span class="dv">0</span>) <span class="co"># 1 = rodent, 0 = other host</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>  boot_bin_mat<span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(bin_mat_adjust[<span class="fu">match</span>(bag<span class="sc">$</span>bags[[i]], <span class="fu">rownames</span>(bin_mat_adjust)), ])</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>  </span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>  <span class="co"># train XGBoost model</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>  model <span class="ot">&lt;-</span> <span class="fu">xgboost</span>(<span class="at">data =</span> <span class="fu">as.matrix</span>(boot_bin_mat),</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>                   <span class="at">label =</span> boot_pheno_mat<span class="sc">$</span>pheno,</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>                   <span class="at">nrounds =</span> <span class="dv">10</span>,</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>                   <span class="at">objective =</span> <span class="st">&quot;binary:logistic&quot;</span>,</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a>                   <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>  importances[[i]] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">xgb.importance</span>(<span class="at">model =</span> model)[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)])</span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>  probs[[i]] <span class="ot">&lt;-</span> <span class="fu">predict</span>(model, bin_mat_adjust)</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>}</span></code></pre></div>
<p>The last step is to calculate a median for all the feature
importances and the number of times it was used in the 10 models.
Additionally, the median of the 10 classification probabilities will
also be calculated. This will allow us to identify potentially
mislabeled strains.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># analyze the classification probabilities</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>medians <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">do.call</span>(cbind, probs), <span class="dv">1</span>, median)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>df_probs <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ids =</span> <span class="fu">rownames</span>(bin_mat_adjust),</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                       <span class="at">probs_med =</span> medians,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                       <span class="at">host =</span> pheno_mat<span class="sc">$</span>pheno)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co"># analyze the feature importances</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>df_importances <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">do.call</span>(rbind, importances))</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>df_importances<span class="sc">$</span>Feature <span class="ot">&lt;-</span> gene_names[<span class="fu">match</span>(df_importances<span class="sc">$</span>Feature, <span class="fu">names</span>(gene_names))]</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">table</span>(df_importances<span class="sc">$</span>Feature))</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>features <span class="ot">&lt;-</span> features[features<span class="sc">$</span>Freq <span class="sc">&gt;</span> <span class="dv">1</span>, ]</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>features<span class="sc">$</span>importance <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">nrow</span>(features))</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(features)) {</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  features<span class="sc">$</span>importance[i] <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">as.numeric</span>(df_importances<span class="sc">$</span>Gain[df_importances<span class="sc">$</span>Feature <span class="sc">==</span> features<span class="sc">$</span>Var1[i]]))</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>}</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>features <span class="ot">&lt;-</span> features[<span class="fu">order</span>(features<span class="sc">$</span>Freq, features<span class="sc">$</span>importance, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="fu">colnames</span>(features) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>, <span class="st">&quot;Frequence&quot;</span>, <span class="st">&quot;Importance&quot;</span>)</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co"># plot the top 15 features along with their frequences in the 10 models and the median of their importances</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="fu">head</span>(features, <span class="dv">15</span>)</span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co">#&gt;                               Gene Frequence  Importance</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="co">#&gt; 29                      group_2194         8 0.337915338</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="co">#&gt; 108                  menC~~~menC_1         8 0.047897755</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="co">#&gt; 26                      group_2082         7 0.046599902</span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="co">#&gt; 17                      group_1477         5 0.082930411</span></span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a><span class="co">#&gt; 25                      group_2077         5 0.033121750</span></span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a><span class="co">#&gt; 24                      group_2016         5 0.009599232</span></span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="co">#&gt; 20                      group_1839         4 0.076863792</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a><span class="co">#&gt; 95                      group_5689         4 0.014925967</span></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a><span class="co">#&gt; 53                      group_3058         4 0.006453217</span></span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a><span class="co">#&gt; 43                      group_2844         3 0.346190933</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a><span class="co">#&gt; 87                      group_5288         3 0.065256526</span></span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a><span class="co">#&gt; 97                       group_858         3 0.052755694</span></span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a><span class="co">#&gt; 4   arlR_2~~~rssB~~~arlR_1~~~regX3         3 0.029352658</span></span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a><span class="co">#&gt; 111         rmlA~~~rmlA_1~~~rmlA_2         3 0.018808632</span></span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a><span class="co">#&gt; 46                      group_2882         3 0.013610958</span></span></code></pre></div>
<p>These results contain some of the experimentally confirmed rodent
colonization factors and we can thus conclude that the models (at least
partially) were able to identify causal features. In the next cell, we
will plot the classification probabilities. This will allow us to
identify strains that might be mislabeled.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>df_probs <span class="ot">&lt;-</span> df_probs[<span class="fu">order</span>(df_probs<span class="sc">$</span>probs_med, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">ggplot</span>(df_probs, <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(df_probs), <span class="at">y =</span> probs_med, <span class="at">color =</span> host)) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;XGBoost results&quot;</span>,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;analysed strains&quot;</span>,</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Classification probability&quot;</span>,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">&quot;Host&quot;</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABUFBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYAsPYAv30zMzM6AAA6ADo6AGY6Ojo6OmY6OpA6ZpA6ZrY6kJA6kLY6kNtNTU1NTWRNTW5NTXlNTY5NXl5NbqtNjshkTU1mAABmADpmAGZmOgBmOjpmOmZmOpBmZgBmZjpmZmZmkJBmkNtmtttmtv9uTU1uTW5uTW9uTY5ubqtuq+SOTU2OTW6OTY6OyP+QOgCQOjqQOmaQZgCQZjqQkDqQkGaQtpCQttuQ27aQ2/+ijk2jpQCrbk2rbm6rbo6ryKur5Mir5P+2ZgC2Zjq2kDq225C229u22/+2/7a2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb27bb/9vb///knV7kq2Pkq27k5Kvk///na/Pr6+v4dm3/tW7/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T////CgigpAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAPRUlEQVR4nO2da3/bthWHaSeZ7SlZmmzOuqsTZ+3aRd0l3aysa9e66tZWXZctCdek3qW2piizo/D7vxvOAUAAFEkQJCjezv+XSCQB0sZj3MFzEESkXAVN/wJtFwGyiABZRIAsIkAWESCLCJBFBMii0oCWoysn8D3bOmafq6++EwTB9b/ClTBA3XhofcTPHhX+casJ+3EuN/hS+RwUBmP2uQj22efFIYeCJwJQsG1LTWiNoYSAXG7wpfKAVhP2614cQD5aTYIbLPOsno4gO/FkrL5AgHnqOSBWyHZEAVsEO+qSTMZyBNnpBctblx9G+tHqD0GwdQ+o8hwHqd8Jg+2HGOW6FoVjgWA4+prfIMM2pCqV9Cx4i6dwZvxlRQ76C7BbjKCwGUcIJgjGBqBLo+DKyTIZJQ1QHLYhVQF0cRBgRX1xsKNfDlV9xFJzDy5cOVFHy9HNEyOrYTSMDBnjKQuJoxiAeBFTt29GlZr5GS9aAhCCgWQIQCohrCDqR5fextZOBwRHvEzC1ThKKiB5+2ZUBVAosnoSEE/3VwFkBSwKC8AijxhW0QnQAAGHhQCrRUkBpG7fjCpV0leeHcCvzDMAyGhrZpAVUgBFTw95JyAbkIySBii+fTOq0syzlIRYyBbBTXktCWi9iGHYs/dY3ksAEgylMApW/5BDzWYewzajKh3FfUEJKlnsRD+7axQxvWpWR4vgeye8jQsFrbgkbX3IegMsO6oos+AX8HQJiF1RYZtRaUAL3oLxEUfck758rFox+CNnN/O8NpfNPD5rIVs/FQUvXT6UgFQzz8c5G1BZQBcH4m8Y8kQ+vQtjsQ/5FdR1DF/GHcWl0VG8ybHuYDoFIIxy6Z4eJfrHKLjxtayD8AYVthHRaN4iAmQRAbKIAFlEgCwiQBYRIIsIkEUEyCICZFFpQGdM+JF+UC7MZ8o8iQBZRIAsIkAWESCLCJBFBMgiAmRRFUCnTAQoS5yPYESA1qUAnRKgNGmA4oxEgDSdrRE6LQZhQIDOojVCBCiWSFaCEAGKJZNFgDJkpo8ArSmZPgKUEAGyKBWQGnwQoExAyS4jAVoDdEqAopTpjqw+9VABrUsD5O+hjcv3hBnlIKmshBIgIQJkUWZCqRXjKg2hw4Did5+LKC8H9beIzYq/wp6V0LWps14BQiM5YR9gUT6g094C4ogKGBtZAJ32FVAYgIHTzF7QMhPaZ0BgScPt3uxZKDuh/QV0ceBgx5eTUGNeqF+AimafiABZNTxAsaFgUMgmPS+h/QQU56Biyk1o/uR0ZwE5aYOAcor+i4dlf/+iigGx3wK6iIX82oBaAWh5tXbj55pG8xohApSVgzLXfyoAgq4stCHoKGYMn4UGjhWkFbG4FfNUxGoAxJ1U7PCcsxyNKQehVN2IpYx9LK/xv2FnAdVVBy3Qx8dVdA+DJW2DgNZbsZfv3vrxN3Dw71ugI/z+4RMRPx9QybD03zEFEGYrlpOazEGvPmFIfiLPnjNWj4+04GYAoRcQ8QdklxsF9PJ3T6LzX4sM8/K3n0ev/vS5FmyDUEsRk5U0MmJ0nPr+5WQCWmj+085/8w1yQUFWYkUOChrTt5ksz/W5Ap3SzC+ET5lZsYFjBRmAQu7OiHctoFBJQPh9/ks9F1lykM9WrFFlT3foOeg5r62Z4nqIAOl10OP7Ms6gAXE0suJ79cl92YrxggXZ6NWfizbzHivpRpUz1OD9IMhEoqSxftDrcUNmA+SxmW9UtDZvEQGyyADEvc16GM33FNBqsr+ajAv2TjcI6CxdHlJfQMlmfrYfLQq94DFUQOFO9XWx3gKKZkgnpBykKfHywj4b/hV7zWyQgFxEgCwaJiB9PsiiQQIy5oMsGiIgX6+/bBaQXAJyVcH7ugVoPp97A1RQ2fNBFjUBaD7XCYnQ5bU3sd7EdaBo9eCDINhf4JI0v4IJ4wNMGGuOo4s770NXBsDyGMvX3uPu09OGofUsPW8U0IiPjvggCRemYe8Slvx42HRx5xEMES5uIxbYj0iszrIY4Y58goiVBchV7QHEEgrJZsljBFYPjiP4zw7FFQB0O+77sotQRFgUuIddZxfkE26n9ZA7BSizDmL/4YslWwMkrmCcER8fzKB4IIjZGPMSrh6JJ8SxcgCFYk+etgIyZALKy0EgKEMs64icInLQnUfaE0SsPECd7AfJ5Ik6SAFSdRB8sv8YE5Zjd+QZjyGeIGLlAHJq5jeoYoBkK6YAqVZsxlueEPboGF/c/rnWiokjxJXbikVd6AelAXJWemWcpfqKmL+Fw/YA8llJ57hg6OxYzElDBLSa+HrTvqeA/JkinLW5DnKTWUk7bEplBVQmLPtHtQKQP1OEM56D+lbEnGQDlOPlhAD1F1Dob9K+l4B896T7VgcNZCzmJgJkkTlpDxNq/uaDNloH1bW4kdYPKtQVsgPKdANTHtB0Os0CVJdqnJP2D2g61QmJUHMV5/Jdz+YtPQBkrOIsfO9xXOeqhvc6KAPQWJuDF8sY/tStZZ+0Okit4ixfOxk6IEMi1FzFIUDrgMxVnFrroJa+SG4BZKzifKvOHAQL/5pio15pzasuRK0CpCPx3l/MnnLVjHq5kZhp5dtCQGCv6buE5UzaK4M6YYdpWPm2B1DNWn+BSkqZZAprXnWhgFGvT7UGkDknrYx6hTWvbuU70BxkyDALh3rIvECAzCqHAaI6yFybV0a9wppXs/KN2jcfVJdy5qSVUa+w5nXsB9UwH9QwIM9TrgRocIB8v6PYvzqok2+51qzuT3fUrJ4CEm+We1A9njjrArS3t9cYIFc1AWhvTyckQsWyj7TckYYJf2OnX14/3H5zHMFcdWl1auk5HRBf9jHfmhf/RuNosQ//hg0Il32EZYYJCKZh3zj5osokWi0Ot+sClFoHqWUfsbhhAlo9+OiNCiWsxrdc6wBkSP7WfNknKwdF4VuV/OF2qpJOB8SXfUQdBLjCbR1QRWeUvVn2kbY9rKb4LhoRribbXwKg1e8rNffd9x9ks01Z3iySnEx133+QBVBYcSGI/AdZRP6DLCL/QRZ1v5mvWR536q1N7QEERSzxikemBpmDZsCmIKEhAmr9aJ4AZf+oVgDqpmuKmpXi5I2WfXRRP8iiXgFKXciouLrRrWWf3d3dPECp8gXIVU0A2t3VCYlQae1z/RAmyHQnb+jTDefRlKMcZ5k5qOVz0umAhLXPCN1LaU7ehE835cxtsIDG2nKG7mApioQLN+HMrRogbdmnW3WQtuyTBIQupZQzt4qAOr7ssw5I+HRTztxKqVOVdDogZe1jAhI+3QyXk+7qVjOfCkhZ+ySKGPfptppgK1Z27r4HOcjJJZmzCJBFBiBezIp1qVoDqGYlZxTD7UfFulRDBMQyULzDolUDBQSzZQRIV+LlBTAanokipkwzzt/B7TGb2oi2NYBg+LsTzUSPSpmogpEYGNU1tRFtewAZUuZhzwHT4yO3jWj7D8g0MGRHThvR+lR7ABn9IMNEFYzpnDai7WcOMvpBeg56+a7caLXoPqtw1DtzKLMfpJmonr8TV88OgISjwH4B0vpBykRV8HHbiLaPgFL7QSwTQf8HqmfHjWibAbTm89nnqobRD7LIDqiGOijeH7khQC4qAMg9LPtHST4aIREK6z1yXQe8vB2OjS17Kqz4oGoF5DsHpQMasSwzU/v7BGNzyx7PRcznlKv3OigdEJj0cEMN9NE1GZtb9ngE5PtNe/+VdFodJLcNESsb0WxsbtnjEZDvN+0304ql5CBjyx7PgLy+ab+RnjTPNsk6yDA+9AWom2/aIwDRVq0mshXTFoIm/lox72/aw1Jxj3rSTioCCF82IEADAVTHJpC9AuSqIoCoDtoMoEalAcIu9GpS8DWRIoA8FbFGpQDB2A4087c23y9Acp5MHeRqcIDUGNXj0nOf6qBaALmG1ZPGSooBwTCDy+dYrEdFjNtCRTqpXBUCZLzY3HVA0QxHqfDiuren81q609I7ijjjWvRt0AHmIDcVAtSrOshRxQA5hvlMmScRIIsIkEVUB1lU+3QHtWJWQLvpYQSIABWrgwgQAUpVQUDT6ZQA5QCaTvd25/M5AcoEhHx2cW9CApQFiPGBXESAUgCdcUBzLuntmABpR3sK0N7efDpntdIeO9/dhTN2OIWjPbGDqs+UedIGBqvzuQSUz4cR8pkyT9rMaH5KgPIBnZ0RIAsgPKA6aFgTZsqoVxw1tRFto8oGpIx6xVFjG9E2qiJGveKosU0gG1URo15x1NhGtI0qG5Ay6hVHjW1E26hK5SAQAbLUQaD1cpZS8gpdam+JzWvF7set2H3eiukb0YKGDUjbiDatHwQaOCC7CBCJANlEgCwiQBYRIIvKA0o0+sLBmeb4DPqUZgcBL2kR/8N9o2X1Ilqh0oDMyQ/pN0ddfQ4UzIkSvKRF/Nf30Tda6mxKa1QaUGLgIVxTxVcfv/4p9DD1QcrHeEmL+N8fIK/MkUwrVBqQOXSNhIMz7SokNjHMFb1yPaIKTTywLSoNyJz8kA7OtKtAIzFRgsyMiDC+S51NaY185SDU4yN7DjIifga+0Xqag9KqjLg+gZPzZB30RAeEl37106MoZzalFarQihmTH8LBmXYVEpuYKJGlTkT8560j7UnrsymtkM9+EDg4K9QPEhF/JHyj9bMfNBQRIIsIkEUEyCICZBEBssgPoOXVXEvXpCX+i4dOtzeqJgC1mceaCJBFjoDQcnyfJfGPsD2edn6MLlHQQWiAIatJIN2rbn0gAPEw+NxfXnsf9v6Kbzef6LCNV91yA4QOPkJI2JUT+I7Prx5DLllNxpg7liP03A1OLsBB9cUBByTD4Bu22Yu02+UTZZxaEltGboD+B04ZMDlj/I7Pr2rbnWHMBc89Y/wWXi9kGAc01h6nnljRLaR/udZBC+4mFf/O8KHOWfECb/izIMDCxv3F7aOnFJlqEYaAeD0U3x4/UcRpjVyL2NZxpCVHP19e+zv3dY1eMaULGQOQCIsBabcr5FFVD9B+5QYI/eAuVA7Sz1eTtyUHKFrCBwgWMb0VY2ExIO12DZDjZos1yxEQ/MVHGiD9PISygSzYKTqLY8GwA6GspGUYVlcISN0eP1HEqSWxZeRYB7EaYuujg7FeY6hzbHsWwoMMNPNbx2YzL8NmwQ5noG6Pn7gov+NnLfI4FmtdA+RFHgGFxTaq75i8AcKeXg9F0x0WESCLCJBFBMgiAmQRAbKIAFn0f1DfRUF2y/Y3AAAAAElFTkSuQmCC" /><!-- --></p>
<p>Assuming that the model does not overfit or underfit we would
conclude that the majority of <em>L. reuteri</em> rodent strains are
autochthonous in the host while a few strains appear to be mislabeled.
For more thorough discussion on <em>L. reuteri</em> adaptation
mechanisms see the main <em>aurora</em> manuscript.</p>
</div>
<div id="example-3" class="section level1">
<h1>Example 3:</h1>
<p>In this last example we will compare two plots. One plot will show a
the original <em>L. reuteri</em> dataset, the other will show the
boostrapped <em>L. reuteri</em> dataset. Using <code>get_bags()</code>
should result in mitigating the bias that originates from population
structure and sampling bias.</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABR1BMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYAsPYAv306AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kJA6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOgBmOjpmOmZmOpBmZgBmZjpmZmZmkJBmkNtmtttmtv9uTW5uTY5ubqtuq+R5eU2OTU2OTW6OTY6OyP+QOgCQOjqQOmaQZgCQZjqQZmaQkDqQkGaQkLaQttuQ27aQ2/+jpQCrbk2rbm6rbo6ryKur5P+2ZgC2Zjq2kDq2kGa2tpC2ttu225C229u22/+2/7a2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb25Db27bb29vb2//b/7bb/9vb///kq27k///na/Pr6+v4dm3/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///80NhuRAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAQ8UlEQVR4nO2da58cRRWHmyW4MwMYTHQTMbAKKIw3lEVQRAaFURGiSUZQ1IXZHSbZTOr7v7budU5duqqv29tT/192uqe6urP9bNWpU/eCZJWquOxfYOjKgCLKgCLKgCLKgCLKgCLKgCLKgCKqA2hVcF17n3/b/fX5onj6DXltU8wrPWt3ckgu7hx6r21/dM8bv1c1AFRwFtuZOL8pri2Lai9QBmh1cFUBiV/88+KZU/pyxc3/EPKvY5FytrNvP/VBlWeVvfBVB0SW9Lgqjvj5dkZpsUufzo5UvN0JC2OvRM8+eb546nUW+jnNkfJsVlz7TKYgE5fs3qUR3qDnNF3SZz08Lorrd2H8XtUQ0O5E/ZW/ZB/0VcWbckFAOk+u8Nm1YxuQjDuXgEQWZslSx+9VjbKYYz6YiTZmGr304Sl9Q3Z2cJelN3brU++ThyeFDWg7u3nKI/D/Z3dCExP5AsXvVU2MNP2zWoBYpjNBEBBjKq98+clvZwUDwfKiQIUBPf2Tz8T/Q28Ssdipid+rGgBihgEDkiWaynXIBikLI+McypS2c23Qkl2/cVcA2ug/honfq5rYIEKIsUH/pK+kktaRuugCoqXed37/t//eKQFEvjgWlMcASJdiF3eYxeb2+eKOMtNLlbFgBhLOk8pi7CpPhkuQCQn59zsUh8hiyqKZ+L2qKSDjBx3Rv7aAtVSu0LJ4ndtnBIhaa1pyP3NKje4b4qoApOJuiu+dkt1f6DNW9IcaaeqwP6QJ1cTvVU0BaU/60IDZqJfYqILZLfDpd32VAwJxVYRVwamrXKtj9KrGgMjDd2eiLqYLGGCYZsWN/2Ejzfy+a+/zDEVtzbV/6KqGiiscRVrUk4tj5hxsj1VVT8XvVUOozfduV6poCIB6922qaACAVtovGKIuH9BuWdw4jUe7LF0+oIErA4ooA4ooA4ooA4ooA4qoDqAz+yQ9IB5zaMqAIsqAIsqAIsqAIsqAIooCevSzB4Q8fvPWi1/roPqAJpNJLObQFAP0za3vPyBPPnyLfPUDHVYX0ESoPObQFAF0//ZHNAU9/vUDkZKEMiAoBubRz78mj3/1sQw5q6mJUjhKxy9bR0mAvnkRAqqbgjSg0phDU40U1BjQpCzm0JQEqF0bNEJATz78RYul2AgBteMHrdejtEEe1QK0XmdAJQFrKYpnXR5zaOoZEFdpzKGpH0DrDKg8IAOKBDh8MiB8kgHFAnIWiwRkQE7A+fk5CrAI7TugcyGrqpEB2YDOMyB/wLkHULZBxAPoXF3IRhqenJcCKr11cOoJkIVn3IDi/ReGjwoxgEpvbP/9GqvjFOQ6ivx83CmoVimm+aiEFLp1aOoWEMmA/AEhPwiUZqFbh6b+AIljBsSVAeGTtMoqP2ZA4ugHZNc3fLcOTb21B/la7kEeLAO0fTawXsrDu5V/98rqcYwi4nOmDZV9i6sQoCC4NtUfoHUGlA6IBVj12VJA74n1YtiMcTa9lU9Fn7PP7udydlUXc2Ng81wJEFubaMVWdjgUixOwlLOdza9qCoLvHQDEPqsAmvP8tGGT7enH9jkxLX9cgED+kscKWewDAYitTsC+LGVOuxKAptMpDkApw3UQ18QCZD/UlQOILxlCU9IVADQVgleigPRZZUBsbZCNXPbi4s6VsEGpgDwJSJihZEeRfigjzRnR75RRjd++otIAfXXrFpuSIGVeajrFhII2yA+IeOyZRwqQLuY3ctm3qos21lEaoPtvwW9pgAIpyOSxCX5W7VfoVkmAnvzxY/g1Aggi8tggLT6ccxyAHr9Js5hORLAbQvIJdGuAEISGfZuwn7H0ajz66ccwFZUbaQLtdKCqMZmo7DaOFMRl7FC5H0QCgAhgtDeAZOKpDsjkNuehA1MSIDYd6smfnGJe5y/nLV0jLQ5uUT8OQMwPum0KMgsQtdH2lQigydhSkCX5LlMgfEU0ZuAAeeTpZ2IIjRjQtBRQIIAd+YSW0QOaWjJX4DEMyJRj9i0DU/+AzMzMyYgB2XzSAHHDhPnsCyBzBR1RgCzaMKCx2qAGgMheAMKE4Fv6Pccz4F6jHLYXgIixQSHPEQLyF2NJgEwLbW9qyQ+SiJIATSZiFRhTuYcxS4Rb+ZVUJ1BVJd7XpE3aBWTbpKCRJnh2y/oSACWqPUBT0jEg1BtgtH3uVd4vzXuCyO7t94riaMM7pUUI00Zs1cA6q+fk4uXfsRZtBlbE2L7wDn/Cxmzo0A4gNwmVAuKvqFeAsUfD1Ac0OyKsx2zJD3LdfP76IoRGuXj5HlkdkouXOBa2KYHsn6UxVofqCTJWp4DKbBAT8KInpC1A9EXZa9PXowR2b39A2A89lSE0ivjkooGst4hGYffQcBqgnvCSr5uteb8YbHatAmhSFVCJDaI/7EBfGwCSITzOjHcTsT7rAwFiOedpifcfySfoWEgt9M1DQK4fJJOW+IoAsTQ0lXymjQGVpSAmlodo0pEpRaagl++BJ8hY1uNrz9WYgr6MqduvEej1wITodQVondSr4fOD1OtJG2QAGRvEPukPj8k6ZA/VNxFDPkHGag4IVDVMAD6xrZQOsJIQi8DwTIFLXVUKkCrFDCBTii1F+bRi+3LML156DZRi8ozjaq8Uw95PGJCOZi6s9SIwEhAXqNZ3Lr8xDqlnQLpRWvIhZ2uBZ1yA7OK9ESBY3xigWgAULtUdG6RpTNSSnNphHFVtvjkgFiBqGHuRgpgl8nWDOX4QAKQSzhgBOXymyoeLNtqbumkGhE4cQIbLCG2Qw+fMrkcuFgtx4nb7yLSytjUmQOcRQAshdurrOOQyZKxp0ENTC4AIqQEIDZxGVwamdgDxIGWDFgtDKJTFYObCMYamGoCcHDZV5X4ckGWZ9w0QSQWEhWKUShv//tQKIEOKR5B8wgOoPIBSPCFg21x11bkxDEA4t4VUCqgrAUCqvWjjthohnYcJqc4LjicVEEkEBLMuEO7FufbDlie4GEBLvnnuvA4gbINg43ESIMdg+xUChHpxNm6zezNpQGLqDG+vdQHhBbdbBmSbo5CCgOagDV52Y7QnCxCbbuQCsjYe8fhBuNLOPUdxogLksQmggA0yvTjbF047BMSzGIH9tUbWov84jxG3bupeUcSa8CkB1EcK0pt9m62steyNRyCg84TOHji5Zc1W6aoHyO8HWb04ndmgMtkbj5yDNMQDiD5hwpmPSRulqD9UXbgX51sdpqCw3I1H+NtOVceYxcMAkn9xU5dVPawundpVDdxJ0bq/CABxMy12hrcU23gkBEjZDAuQ3S/fFiA2Y7PtHAYAsUEgTEu3lI9sPGJSzBn8Nl2EAdmErkJldXlonxiVbzwCAcHybWETEnc4QzvwQwcm2w8icU86DGiK+lwXFQDVNtGdqw1AoAMVFm6o2ePczP6xAYGm/AFKA9qdqKVmVh4zjRQGNIWAgk2vVxMQW3+GHw2pkNxOneqAJgaPtXDOwOR40qxyHLknAZC8EOr+QUnoygASCzslOBKebsEAoJJeDdAIdGUApaoMEDVINNkEe1Z9gNb7AMg407jKXQlQjV+le3maXKNqB9DVSEBWe9AqiZA/AwlP2mr2SwSU6igC69aXgB80lx9RBS0MBLTwxNDjyhxASVUNqwAokxxZ3oKwJx33gZhqAjJDW4llg9AzwhoFIItQm4AcH0L+4qLbR83cURMT/k6/fnr9+ODVOWFt1bXVISCGCMSAo6MxIPyMoEKARLcPHjUv/83mZHPE/tVX64AQoUUSoEQbFATEu33kzAwMiDXDvnL65yaNaABQoVS5No8DagBKm+3jt0Gm20d2bmBAu7f/8EqDHNbFxiM4CRnpGSxKEFDKEl1hQCUpiKx+3GhF3LY8aXSCzJC+YBtpZ4HpuHxlmOz2kTaI4VodQEANl6PsBBA58wGy/CCCCTVotH8NlGJsRs93+STC3cnBpwzQ7jeNivv2AVEqU90eXVYXaw1QeQrZ3qz7ZK62AUkwSYCqGmm/IoBWDTuCugLE01B5bb4lQB2rZUDa9kzx+nhRQKMaSJ4CiCHyxAi3B40fEG/uSAe0tlXjF+lDLbcoQh+xDJCDZ18AQUJWDHTcO0DTaoA8fPYFEECEY6DjXgMCw17igMg4J/XaJ2ckMOwlZoNIBuTG5AKARjihzusH+Y1QqR8kA9apvRqoOaAvteoousVYeW2egA76BECoQcknb0dGw96NlgFBRzEBECjD2gDkVS+AQhuP4KMFyBMjDCjBSONGbS012+f6MWsgg4u88TXdeDuab/pAotIAhTYesQPsBBQBtG4FkJztM+PLS4FF3uSabmYxt1pKAhTceMQOsPD0BGgOujPgAkuEyCXc5GJuKa/qKAlQeOORYIcGDApGR4DivRp+GwS6fWxAfLyKWcytMhymtKkIwY1H7AArAenBMG3YoBJA3hQk13Qzi7nVUhTQ/Vu3xBj7wMYjOMDi0y4gvx8EZvtgQHJNN7TkZHWlF/NJgIg3AbmN9zUB+QRm+1hZTKzpxqZ3icXcail5OpRn4xF/uoAJJgaItAKo033qkv0gz8YjYf/YAWQPyfNV62tqEICw0gE55X4HgDpWBhRRB4AWaJy037P2t3sMUa0DcmxOBtQIkF6za18AeepiZTbIDRiaOgUkDhlQAJABE/aD9g8QxqJSDhp5F7x1kOoOkElLoBaeARHtB2VAEU9aA4ItgQ0Brfvvoe6wqmEloOaAtFNZJmfC0mC6fZyADChaWZWlWUs2yFRskVh/j+rXYau8Hc/Rlj0Neny4Oq7Nc2PUMaAZTTJLs79PMcdb9vSfgmJzNTw9HRRPSeeGUen/GwLEpvSIiRp8ja6TOd6yZ8hZLKG5o7kNUtuGyJ4NspzjLXsyIDcFoS17Bg4I8+nCDxLJxrZBaPJhyqNDulqAfOIAZFm1O1GlGOgIclf1q6LeAKXeOjT1ZYOSbx2aMqCIuvakJZ40l3KI6r4utvDNG8uACAQ0yYA8AQsgPWolAzLHDKgSoNxo7wYgPhmQG5ABpQICIy8zIHDMgNIBjdkP4iuRl68nHQrYC0fxGzaJxdpXIwMyun/7I5qCYmva73VllYGx99VI786QSr2l45etoyRA9r4a9okvADrS40xBcpKGm4KqAxp1c8ejejZoMf4UJMXARPbVyIBq+EH7A8ijbIMiyoAiSq+LsUEdvDkxA/IGmLFlGZA3IAOKBGRAkYAMKBKQAZUHTDKg8oAMKAPyq7INyo5iICADigbYI+wzoEhABhQJ2AtAqV0UNdT++zVWTkERZUAR1QG0V8qAIsqAIsqAIsqAIsqAIsqAIsqAIsqAIqoOiI9mgCtwu8Mb8Prc1tAH97pvgMRgVBkQH9UJV+D2DPPE63NbQ0Cd696BooNRVUBiVCdYO9kzzNNan9safuVc9w4UHYzqZTG0ArczSA+vz20P4HOue4f5DUb1AKEVuJ1hntb63NYQUOe6d6DoYFQBEBjVKb7/0jPM07M+tz9tQDs0vhTEpV7RM8yTQAB+62IDGpcNQitwO8M8rfW5rSGgznXvQNHBqL4fpFfg9vpBt7FRdvwgeH1cftC+KQOKKAOKKAOKKAOKKAOKqHVAu5Oi0FsNrcymQx1vf9GZOgB0xA6rA7YgHdtPZ1XwtUOXDfd9vyx1BYhtzbQU+w2t6KHB5jqXrA4BMUZKm8PtsxkQlwS0PLiHkWRAUtJIH9yz1inOgKRkCiI2kgxIygBSNkgU8BmQlAHEiy8iSvwMSAsAQn5QBiQFAPF9PJX/kwGNVBlQRBlQRBlQRBlQRBlQRBlQRBlQRBlQRBlQRP8HgtV3phQThQMAAAAASUVORK5CYII=" width="50%" /><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABQVBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYAsPYAv306AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kJA6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOgBmOjpmOmZmOpBmZgBmZjpmZmZmkJBmkNtmtttmtv9uTU1uTW5uTY5ubqtuq+SOTU2OTW6OTY6OyP+QOgCQOjqQOmaQZgCQZjqQZmaQkDqQkGaQkLaQttuQ27aQ2/+jpQCrbk2ryKur5P+2ZgC2Zjq2kDq2kGa2tpC2ttu225C229u22/+2/7a2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb25Db27bb29vb2//b/7bb/9vb///kq27k///na/Pr6+v4dm3/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///+9E9ujAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAO00lEQVR4nO2djZ8bRRnHt0fxkhQttnoH2HIIKgQVtFdBsRIgvpXqtQEU9eCSNL1rmP//D3DeX3beNrs7m8nm+X16l8nk2bT77ewzs/PsPFMgUFDFtv8BuQsARQSAIgJAEQGgiABQRAAoIgAUUR1As4Lq+gP6bv3X7xfFC+/wz5bF2Hvc+vQQXd05dH62+umZ037ragCooCxWI1a+zT6bFv6TCgGaHfQJEDuZL4sXz/EJF7f/g9C/TljLWY1+cO1j33GhE+4jIDTFr7PiiJZXI0yLfPRwdKQsv8RX37W3aWlUXP+ct6D1KbGlp7/+EBu8g8u4DeLjnp0Uxc1Huv3W1RDQ+lT8z39NfuHTZ2fPDeWVSEvXT8qAKBhswAGxy5U0QWm/dTW6xCyXQly0ctOY3iPStojZtQfo2WlRBrQa3T6nBvQ716e4MaGvDPutq4mTxv/VJUDkojOqvv7b70cFAUGuO4bKBPTCLz5n34mPZFakqOy3rgaAiLMwAfEeTfpb/v6Qt6q17YOm5PNbjxigpQSv7LeuJj4I0auIl7/ApymaFnfTuIf74R///t87AUDoqxNGtK+AZC92dYd4bOqfr+5wN70asYGSuMQIHNrkyJWoGt+/P8A42CUmvJey37qaAlLjoCPcAhisKR8KYTTnpOd+8Rw7XdqXC0DF2+zdsvjxOVr/BdvP8A920nhw/gw3SmW/dTUFJEfShwrMkp8Y78TJeHKpd/PqHbfABrOCEhZXqLTYuhoDQs8+HLF7MdnpSMdExn3XH9ALCvua6/+UtxpfjIpb/5MDRdzVo6uTAje31Ym4rRP2W1fXd/NZ+JVN1DWgLMY2m6hjQDM5BtgVdQtoPS1uncfNchLMKEYEgCICQBEBoIgAUEQAKKImgOaOUtPK7ASAIgJAEQGgiABQRAAoom0BGgwGAMhfOWByWGYnABRRFNDlr54g9Pz949e+tT6qDWgwMAntMqCnx68+Qd99cg998xPrMwCE0ONXPsUt6Plvn7CWZAoAEREwl7/+Fj3/zWe8Zt5YElD5g8QnW0eVAD19TQckBU6ayG5BUgCI6DKBD0J9GigSMN998l6rvVigMjttZRwUqMxOcLMa0XYAXVxcACB/5QUTAPJVAqBw5YWQyzI7dQzoQpfLMjt1CuiiJIdldmoF0HA4tCsBkDivIZNZiRyAynwAUKno49NrQEMlVVku7S+g4TAKaLFYeHyQYbmgdhkqMaAFkwOQYcnNckRUFxA5m/kwBmhhAtIR9RsQP5/KgOSZuwEtLLt81CIghkgHZJ/4ngBahAHJUgAQ2gNAZTwM0VyV1Hm7ejEGiE7d980HedpPDJA9kMY/LDzWR0AuPISLKmkXjjYLZInFNzLF0w0gZAJCFqBB9K/cnmoACvHxA/I1IEqo/fNqTS0DWqAJL02kD0JNAa1uePKlPHu0+b9+UyUGRKvigMKIfIC84NpUDUBzP6DFfD6fEESTCS4JI/bkRpDPIPR0x64BQiEXhDAgKgJSVSLHZIcBKNSEVjc+YjlkyCpysuSVLkUfk9/p13fWvNUIAEKI4tGNHPNB+O18oPCEAZHcRDOS2eGQpUQhLWc1GmfbghBCQUBcFiC00PmIZ2Auoj09TVmBaSzJAnz8a/USW5afMyAPIbQxIHGdhf4u1mIwIJIThLyZ8ittxwGVfBBRS4BoyhDcknIGZCJSt/BhQMjkY/okryQgkhtkydNeXN3J2gcROR2PLNL6xbA0SaQDMvu0gAQg4aQpI/weM6r/r6+qRIDUJadVEikemwOS3fySp4ILJW1sS+0AUix4qQIgg88Q5aomUQ1BQdEo1VjRRL3FVG5A21UWgIb5NqDmcTH9UnP2+8ySymwyks6wv5FVKgcgjw8yLylVBEBOQHMknE/PATl8EDKfGSr7IPmm9HGOSgbIOXp0XmD9B2SPgxbaIx2mpWBysV+AaGkoblbNaKHDstSd7QcgcXV5AImwFwCa2490SGTimPIgeh8AyR4eAClVBuSLse4pIN0Hsfl7AORx0iICFAxCo/0DJMdBMkQWCkKjDQBt4RGQli8xo3IykU2IkGgMyPL9VCIItKkqHtcJoBKLPABVVDVA0UW96j4+HSDH8IFo9dJbNC5NI0Foff+jojha0qA0qyFasq0aSLB6jK5e/wOZ0SZgmcXq5Q/oNyy1DR2kKgGKJzcxAYmbjhIf+Whru4BGR4hEzKb0he3WcUhPn9Vgk6vXz9DsEF3dpVjIpgQ8PostZofiG7hVHUDxxAI6IFVyAyr5oIvGLeiMnjY+PUxgff9jRH5wkdcQQHdlAA1XkmgRNiHH4HpcIb5BWW0KqEJyEzUOmssCfRZmoj/44nwOpnJyE78Pwj/kBZ+2BojXUJsRDRORmPUBAzEd07ZE40f8G6TV5oAqJDdRgPQhI5dxNZX5VG5BYUChFkREriHcdHhL4S3o9TPtG7hV6etrtCANi1YaWqoICFUG5B4HidPjPkgBUj6I/MY/1JIEZA/FO2bBv4Fblb6+JR8UBoTmnERTQC4JQKIXU4BULzZl/dOM7NUxvrr7M60X4yWKq0EvFk9uYvFxzyhaHsj8uAu5nbFPLY2DkE3ICchyQD0B5FYIkPFxAFDpi/JTm+vmTT6BqAYAqgrI+qL81G5UQ+dTAZD9RfmpW0D2fWrZMjulAGR97AHk+KL8lMAH2R+LIgczAEClj8uA6POt1uF5qt30OMYNhgVIXlsDz+E5qsv8Qd7xMwACQNUqTUI1APF1RF1qlwDJMJtLqYIb3SZ5SwgolTRAYr5oac8audUAkNcyIBWHNGRGca6/0fICFwVoSjfUHScFhFIAMqI4S3vavZkkILZ0hs7XpgfktwzIC2iszcHzMEZ7KgEiy41SAqJTriHLkNw+SEVxVi+fJwRELzGkx2ujqgUoUhmSH1AXLUhu9i23so6qa0DucVApipPMB9VQHRbOVP+NRtJmFOd7CVvQxtockLFGvjVAOpLWx4saIOqm16dWbNGr3ACRFZttX2EaIPIQCNG0qgvaHJCZZmHXblanh+VCTPsFSK2xjo2DEuzpswt7+1QHJJWHD0osCWh9KlLNzKq66f0CRPLP0FdFKqY8xkGJZY2kyc1xxWO7H0lvQfpAkSZ22mAgsXeANhUAimjfADkf0QtpzwBNWSa1DY7tHtBw2HmWD20cNOa/KqtrQKXYf0j8yfIWZI6kq4+BiABQRB0DKj9/xMXDPmLljliY8A/89uHNk4O3xojMVddWDwCxsI/51Dz/Mxqj5RH5U199AETDPnxlhgmITMO+ef7nJpNoGqBCKNdJe7cPUmEfHtwwAa3v/+nNBlfYTg0U/YACLQjNft4oI+4uAXKPg3jYh/sggmt2oANqmI5ytwC5xMM+IuI5K4of0UWE69ODhwTQ+neNuvs+AAq3kNXtJt/ef0CzhoGg3QeUWAAoooSAtCcNAJBdaTyrAoDsSgAUPm3zeTkAZFXuDyC6Xr7KqmezODEI9RjQ0+NXn1TI/mJXTnRE/QX0+JVPcQuqknmhXNwTQOwSq5D9xZZGqPLzMYlPto4qAaqQ/cUu6k2orRa0hb3a/IAeHx8Tt2O3IKnYabcNKLpNkjOQ0TC6UakF1fFBpNSyD4rvI+VSF4CqZH+xi207ac9OW2K1z80TMkGmJ3mjOd3oPNoGywfKSjUOQsbNRlJAfLXPiKaX0pK88ZxuKplbLaW7m9dHimkBjbVwhp5gCSGewo0nc6v0t5S1Q4A8PkgL+5QB0edVVDK3an9LSanv5rsB5GxBPKebSuZWS+kBRS0bjoO01T4mIJ7TzUg5ubmS380nv1nVVvuULjGW040s72LJ3Gpp96c7NktJtrEAUEQw5RoRAIooESCCBgB5T3uiK2hpVWan9IDaGyhuRSkATQAQU0VAEz0KDYAcgNwjoj0G5CHksrQOz04dApo4LK3Ds1O6cRAAqjof1CIgPX1nR0o/3dGeDypnOHXKWrCUOqoRUNX5IAAUne4gr2w2sAEgO4swFYn3iLgOyfJ2Mja27GkQ8aHqABD+NZfzyQkAjXCTmar9fYqxuWVP7i2IKyEgsqSHLdSgObpOx+aWPbsBSKU1iRwektsHiW1DeGQDTcfmlj3dA9r46RcDUIPHX/yASi3I2LIn0xaEErQg9ziINZuyDzIWH1Y/JVupHyQnaFryQR5RALyvWp+KXkwLBFXO6udU4qUInIsENIFbDQ8g0ZQAkFGpHsgwB47+w7MTAIqoI0ATAOSutPhoiAAQAkB2yao0HJBBCADJ0mTickMASJZKF1ro8Oy0BUDBw7NTF4BKnX3w8OzUKSAz5ToAYi8DpYhlluoSUMwyS6UGNBkAIFQR0CRsmakSA5oAoFLJAiQJwcMLrkodUNgyVyX3QROjAQEgv5MGQO5K6ObNUggQQQSA9gzQ5bvHx/dqrnq278WSRFYTKwyIpFu4/OVndbK/7Aegp4TK43t1My9w9RgQEW5FtbK/6A/CiG2PYqZpz7WWooBIWopa2V94yXj8pVdOmmV/ef7+e6hu9hf6W0xE93QcdPku7sNqZ39B9u6xPQPE+NTM/rIPgL45JrrXYBzUc0BhbeSD4pZ5CgBF1EFczHjwDgA1rMxOACgiABQRAIoIAEXUBNBeCABFBIAiAkARAaCIAFBEACgiABQRAIqoLiBPyNWVfdo1HRnIUp2ZagLyhFxdu3C4wrKh3ToyU01A7pCrcxcOR0gkuFtHZmrgg6yQK3LuoeAMqgX2WshM9QE5Q66OXTicYdnAbh2ZqQYgZ8jVvwvHnrYgT8jVsQuH09UEduvITDUB+UKujl04nGHZwG4dmakmIF/IFcZBeycAFBEAiggARQSAIgJAEbUOaH1aFHKroZnadCjx9hfJlADQEXmZHZCEdGQ/nVlBc4dOG+77vi2lAkS2Zpqy/YZm+KXB5jpbVkJAhJHQ8nB1AwBRcUDTgzMTCQDi4k764KyUpxgAcfEWhMpIABCXAiR8EOvgARCXAkS7L8R6fAAkpQEyxkEAiEsDRPfxFOMfANRTAaCIAFBEACgiABQRAIoIAEUEgCICQBEBoIj+D7vK6C1QdcOEAAAAAElFTkSuQmCC" width="50%" /></p>
<p>The unadjusted plot is not as well separated as the adjusted plot.
Specifically, the porcine strains are intermixed with poultry strains in
the unadjusted plot. As we show in the main <em>aurora</em> paper,
isolates from these two hosts are actually very distinct from each other
but the unadjusted analysis does not communicate this well.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
