<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>outputs</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">outputs</h1>



<div id="outputs" class="section level2">
<h2>Outputs</h2>
<p>This article details all outputs from two main functions of the
<em>aurora</em> package <code>aurora_pheno()</code> and
<code>aurora_GWAS()</code> and a minor function <code>get_bags()</code>.
The output from both functions depends on what parameters were used and
thus not all objects showed here may be part of your output.
Additionally, some output objects are not stored as R objects but are
simply saved to a directory that is specified in the function call with
an argument <code>save_dir</code>.</p>
</div>
<div id="outputs-of-function-aurora_pheno" class="section level2">
<h2>Outputs of function <code>aurora_pheno()</code></h2>
<p><code>call</code>: <code>aurora_pheno()</code> function call with all
used arguments.</p>
<p><code>bin_mat</code>: The supplied binary matrix before any
filtering.</p>
<p><code>phenotypes</code>: All analysed phenotype classes.</p>
<p>#If Random Forest was used
<code>$results$results_random_forest$p_val_mat</code>: Matrix that shows
results of pairwise Kolmogorov-Smirnov tests. This matrix is used to
indicate if the species has evolved adaptation strategy towards
phenotype classes. If both p-values (Class_A vs Class_B, Class_B vs
Class_A) are above 0.05 then the two classes should be considered
indistinguishable. This has no effect on subsequent run on
<code>aurora_GWAS()</code>. If any two classes are indistinguishable
then <code>aurora_GWAS()</code> prints only a warning and proceeds to
remove mislabeled strains rather then all strains from the two
classes.</p>
<p><code>$results$results_random_forest$auto_allo_results</code>: This
data frame contains the raw results of <code>aurora_pheno()</code>. It
shows the predicted and observed phenotype class for each strain. It
also shows a detailed breakdown of each strains adaptation capability.
There are four possible labels that characterize adaptation of a strain
to a class: FALSE, TRUE, TRUE not typical and INCONCLUSIVE. FALSE simply
states that the strain is not adapted to the class. TRUE shows the
opposite. TRUE not typical points out that the strain may be adapted to
the class but likely does not possess some key adaptation factors (i.e.,
weakly autochthonous strain). INCONCLUSIVE then states that neither of
the previous labels could be identified. If the predicted phenotype
class is not the same as the observed then <code>aurora_GWAS()</code>
removes these strains.</p>
<p><code>$results$results_random_forest$plot</code>: Plot showing a
percentage distribution of <code>auto_allo_results</code>.</p>
<p><code>$results$results_random_forest$distance_heatmap_median</code>:
Pairwise distance matrix containing Random Forest proximities calculated
from Random forest models constructed in Outlier Calculation Phase. The
final value is a median of all the proximities. This matrix is clustered
and saved as <code>aurora_heatmap_median_random_forest_[date]</code> if
<code>write_data</code> is set to <code>TRUE</code> and
<code>save_dir</code> provided.</p>
<p><code>$results$results_random_forest$distance_heatmap_sum</code>:
Same as above but the final value is a sum of all the proximities. This
matrix is clustered and saved as
<code>aurora_heatmap_sum_random_forest_[date]</code> if
<code>write_data</code> is set to <code>TRUE</code> and
<code>save_dir</code> provided.</p>
<p><code>$results$results_random_forest$legend_heatmap</code>: Legend
describing row and column colors in
<code>aurora_heatmap_median_random_forest_[date]</code>
and<code>aurora_heatmap_sum_random_forest_[date]</code>.</p>
<p><code>$results$results_random_forest$gfs_importance</code>: This data
frame shows the feature importances calculated from Random Forest models
constructed in Outlier Calculation Phase.
<code>MeanDecreaseAccuracy_median</code> indicates the median of
MeanDecreaseAccuracy values, likewise MeanDecreaseGini_median shows mean
of all MeanDecreaseGini. See
<code>randomForest::randomForest::randomForest()</code> for more
information about these values. The subsequent columns show the presence
pattern of the feature in all phenotype classes in this format: the
feature is present|all strains in the class.</p>
<p><code>$results$results_random_forest$aucs</code>: Area under
receiver-operating characteristic curve (ROC curve) for all Random
Forest models constructed in Outlier Calculation Phase.</p>
<p>#If AdaBoost was used
<code>$results$results_adaboost$p_val_mat</code>: Same as
<code>$results$results_random_forest$p_val_mat</code> above.</p>
<p><code>$results$results_adaboost$auto_allo_results</code>: Same as
<code>$results$results_random_forest$auto_allo_results</code> above.</p>
<p><code>$results$results_adaboost$plot</code>: Same as
<code>$results$results_random_forest$plot</code> above.</p>
<p><code>$results$results_adaboost$gfs_importance</code>: Same as
<code>$results$results_random_forest$gfs_importance</code> above. See <a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html">AdaBoost
documentation</a> to find more information about AdaBoost feature
importances.</p>
<p><code>$results$results_adaboost$aucs</code>: Same as
<code>$results$results_random_forest$aucs</code> above.</p>
<p>#If Log regression was used
<code>$results$results_log_reg$p_val_mat</code>: Same as
<code>$results$results_random_forest$p_val_mat</code> above.</p>
<p><code>$results$results_log_reg$auto_allo_results</code>: Same as
<code>$results$results_random_forest$auto_allo_results</code> above.</p>
<p><code>$results$results_log_reg$plot</code>: Same as
<code>$results$results_random_forest$plot</code> above.</p>
<p><code>$results$results_log_reg$gfs_importance</code>: Same as
<code>$results$results_random_forest$gfs_importance</code> above. See <a href="https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html">Log
regression documentation</a> to find more information about Log
regression feature importances.</p>
<p><code>$results$results_log_reg$aucs</code>: Same as
<code>$results$results_random_forest$aucs</code> above.</p>
<p>#If CARTx was used <code>$results$results_CART$p_val_mat</code>: Same
as <code>$results$results_random_forest$p_val_mat</code> above.</p>
<p><code>$results$results_CART$auto_allo_results</code>: Same as
<code>$results$results_random_forest$auto_allo_results</code> above.</p>
<p><code>$results$results_CART$plot</code>: Same as
<code>$results$results_random_forest$plot</code> above.</p>
<p><code>$results$results_CART$gfs_importance</code>: Same as
<code>$results$results_random_forest$gfs_importance</code> above.
Feature importances in CART models are sum of the goodness of split
measures for each split in the tree. See <a href="https://cran.r-project.org/web/packages/rpart/vignettes/">rpart
vignettes</a> to find more information about CART feature
importances.</p>
<p><code>$results$results_CART$aucs</code>: Same as
<code>$results$results_random_forest$aucs</code> above.</p>
<p><code>$results$results_CART$distance_heatmap</code>: Pairwise
distance matrix containing CART proximities calculated from CART models
constructed in Outlier Calculation Phase. The proximity calculation is
described in Figure S5 in the <em>aurora</em> article. This matrix is
clustered and saved as <code>aurora_heatmap_CART_[date]</code> if
<code>write_data</code> is set to <code>TRUE</code> and
<code>save_dir</code> provided.</p>
<p><code>$results$results_CART$legend_heatmap</code>: Legend describing
row and column colors in <code>aurora_heatmap_CART_[date]</code>
<code>$results$results_CART$complexity</code>: This data frame shows the
AUC for every pair of phenotype classes as well as the most common,
maximum and minimal number of splits (features) used to classify the two
classes. Higher values indicate that the two classes are difficult to
separate.</p>
</div>
<div id="outputs-of-function-aurora_gwas" class="section level2">
<h2>Outputs of function <code>aurora_GWAS()</code></h2>
<p><code>GWAS_results</code>: This data frame is the main output of the
function. If Panaroo/Roary/PIRATE were used as an imput then the second
column will contain the annotations. The next columns show the frequency
of the variants in the phenotype categories. Next column contains
p-values from chi-square test for each variant. Keep in min that the
lowest possible p-value is 0.0005. These values do not serve as a
genotype-phenotype association metric especially not when multiple
phenotype classes are analysed. Next, four values for each phenotypic
class are calculated; F1 Score: harmonic mean of precision and recall,
Recall: true positive rate, Precision: accuracy of positive predictions,
Standardized Residuals: Deviation from expected value. Only F1 Score and
Standardized Residuals should be used as association metrics. See
<code>panGWAS</code> vignette for more information of how to post
process these values.</p>
<p><code>removed_strains</code>: Data frame that indicates which strains
were removed prior to GWAS analysis.</p>
<p><code>bagging_results</code>: Data frame that shows how many times
each strain was sampled by either <code>phylogenetic_walk</code> or
<code>random_walk</code>.</p>
</div>
<div id="if-other-mgwas-tools-are-used" class="section level1">
<h1>If other mGWAS tools are used</h1>
<p>If results from <code>aurora_pheno()</code> are provided then prior
to running any of these tools the mislabeled strains are removed.
Otherwise not strains are removed.</p>
<p><code>results_hogwash</code>: Result from running <a href="https://github.com/katiesaund/hogwash">Hogwash</a>.</p>
<p><code>result_treeWAS</code>: Raw results from running <a href="https://github.com/caitiecollins/treeWAS">TreeWAS</a>.</p>
<p><code>result_treeWAS_ordered</code>: Ordered results from running
TreeWAS. The p-values are converted to standard scale.</p>
<p>If <code>write_data</code> is set to <code>TRUE</code> and
<code>save_dir</code> provided then the TreeWAS results are also written
<code>treeWAS_results_[date]</code>. If requested then binary matrices,
phylogenetic trees and phenotypic matrices are written for <a href="https://github.com/AdmiralenOla/Scoary">Scoary</a> and <a href="https://github.com/mgalardini/pyseer">Pyseer</a>. These files are
correctly formatted can be used to run these tools.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
