<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>SNP_and_kmer_GWAS</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SNP_and_kmer_GWAS</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette demonstrates how to configure <em>aurora</em> for k-mer
and SNPs targeted mGWAS. Unlike pangenome GWAS, k-mer/SNPs mGWAS
involves analyzing hundreds of thousands or even millions of variants.
Many of these variants are not causal. With such a vast number of
variants, <em>aurora</em> may struggle to complete the analysis in a
reasonable timeframe. In this vignette, we illustrate which
hyperparameters should be adjusted to reduce computational time and how
to interpret results of SNPs targeted analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(aurora)</span></code></pre></div>
</div>
<div id="workflow" class="section level2">
<h2>Workflow</h2>
<p>Let’s first load some data. We will investigate association between
core-genome SNPs and penicillin resistance of <em>Neisseria
meningitidis</em> serogroup B. See the <em>aurora</em> paper for more
details. Unlike in pangenome GWAS, the input binary matrix is not a
dataframe; instead, it is a system path leading to a VCF file.
<em>aurora</em> assumes that the VCF file has these columns: #CHROM,
POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT and a column for each
analysed strain.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;tree_neisseria&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;pheno_mat_neisseria&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>bin_mat_snps <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">system.file</span>(<span class="at">package=</span><span class="st">&quot;aurora&quot;</span>), <span class="st">&quot;.inst/extdata/neisseria_snps.vcf&quot;</span>, <span class="at">sep=</span><span class="st">&quot;/&quot;</span>)</span></code></pre></div>
<p>Running SNPs targeted GWAS has a few caveats. First, you need to
decide what SNPs you want to analyse. There are two options: “biallelic”
and “all_alleles”. Setting <code>which_snps = &quot;biallelic&quot;</code> will
remove all SNPs that have more than one alternative allele. Setting
<code>which_snps = &quot;all_alleles&quot;</code> will create a new column for
every alternative allele. A simple way to remove most non-causal SNPs is
a frequency filter. We thus set <code>low_perc_cutoff = 5</code> and
<code>upp_perc_cutoff = 95</code>. This will remove rare SNPs (present
in less than 5% of all strains) and the most common features (present in
more than 95% of all strains). We will also set <code>run_chisq</code>
to TRUE and <code>cutoff_chisq</code> to 0.1. This will run chi-square
test with each SNP. If the p-value of the test is higher than 0.1
(<em>e.g.,</em> the feature is uniformly distributed) then the SNP is
removed. Another way to reduce the amount of analyzed variants is to
group correlated SNPs. By default <em>aurora</em> uses Hamming filter to
do this. Hamming filter however is more likely to collapse features with
lower frequency. It thus might be beneficial to use Jaccard filter
instead. Jaccard filter utilizes DBSCAN to identify the correlated
variants. DBSCAN is an algorithm that may require fine tuning
(parameters <code>minPts_val</code> and <code>eps_val</code>). The user
thus may need to experiment with the parameters to find those that do
not remove too many variants. In this case, the initial filtering
removed 46858 variants, the chi-square filter removed additional 53224,
Ancestral reconstruction filter removed 303 variants and Jaccard filter
with the default parameters collapsed 2377 variants. The number of
variants was reduced to 7160.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">aurora_pheno</span>(<span class="at">pheno_mat =</span> pheno_mat,</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                        <span class="at">bin_mat_snps =</span> bin_mat_snps,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>                        <span class="at">which_snps =</span> <span class="st">&quot;all_alleles&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                        <span class="at">tree =</span> tree,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                        <span class="at">low_perc_cutoff =</span> <span class="dv">5</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>                        <span class="at">upp_perc_cutoff =</span> <span class="dv">95</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>                        <span class="at">jaccard_filter =</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                        <span class="at">run_chisq =</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                        <span class="at">cutoff_chisq =</span> <span class="fl">0.1</span>,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>                        <span class="at">condaenv_path =</span> <span class="st">&quot;/path/to/conda/env&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                        <span class="at">write_data =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The function above will not be run because it takes a few hours to
finish. Let’s load the precalculated results and run
<code>aurora_GWAS</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;aurora_pheno_results_neisseria&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">aurora_GWAS</span>(<span class="at">bin_mat_snps =</span> bin_mat_snps,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>                   <span class="at">pheno_mat =</span> pheno_mat,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>                   <span class="at">tree =</span> tree,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>                   <span class="at">aurora_results =</span> results,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>                   <span class="at">mode =</span> <span class="st">&quot;consensus&quot;</span>,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>                   <span class="at">write_data =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The results are the same as in pangenome GWAS except unique
identifiers. The identifier of SNP is a combination of the contig
identifier (<em>i.e.,</em> gnl_funcentre_CHJKMJOA_1) and the position in
the contig (<em>i.e.,</em> 10560). Let’s print the top 5 SNPs associated
with the resistant phenotype.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>res<span class="sc">$</span>GWAS_results <span class="ot">&lt;-</span> res<span class="sc">$</span>GWAS_results[<span class="fu">order</span>(res<span class="sc">$</span>GWAS_results<span class="sc">$</span><span class="st">`</span><span class="at">resistant std_residual</span><span class="st">`</span>, <span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">head</span>(res<span class="sc">$</span>GWAS_results[,<span class="fu">c</span>(<span class="dv">9</span><span class="sc">:</span><span class="dv">12</span>)], <span class="dv">5</span>) </span></code></pre></div>
<p>These are all in the same contig and around the same locus. In the
next step we could construct a Manhattan plot using for example <a href="https://jameshadfield.github.io/phandango/#/">Phandango</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
