<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>get_bags • aurora</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="get_bags">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">aurora</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/get_bags.html">get_bags</a></li>
    <li><a class="dropdown-item" href="../articles/outputs.html">outputs</a></li>
    <li><a class="dropdown-item" href="../articles/panGWAS.html">panGWAS</a></li>
    <li><a class="dropdown-item" href="../articles/panGWAS_with_large_dataset.html">panGWAS_with_large_dataset</a></li>
    <li><a class="dropdown-item" href="../articles/setup_conda.html">setup_conda</a></li>
    <li><a class="dropdown-item" href="../articles/SNP_and_kmer_GWAS.html">SNP_and_kmer_GWAS</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DalimilBujdos/aurora/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>get_bags</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/DalimilBujdos/aurora/blob/HEAD/vignettes/get_bags.Rmd" class="external-link"><code>vignettes/get_bags.Rmd</code></a></small>
      <div class="d-none name"><code>get_bags.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>In this vignette we will explore the function
<code><a href="../reference/get_bags.html">get_bags()</a></code>. This function is used internally in both
<code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> and <code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code> but can be
also accessed directly upon installing <em>aurora</em>. The function
takes a dataset of microbial strains and returns a bootstrapped dataset
that is adjusted for population structure. <code><a href="../reference/get_bags.html">get_bags()</a></code> has
many possible uses and we will demonstrate a few here.</p>
</div>
<div class="section level3">
<h3 id="description-of-function-get_bags">Description of function <code>get_bags()</code><a class="anchor" aria-label="anchor" href="#description-of-function-get_bags"></a>
</h3>
<p>This function requires two inputs. The phenotype matrix
(<code>pheno_mat</code>) which has to be in the same format as the one
used in <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> and <code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code> and a
phylogenetic reconstruction. This can be either a phylogenetic tree
loaded as a phylo object or a distance matrix that contains phylogenetic
pairwise distances (core genome amino acid identities (cAAI) or average
nucleotide identities (ANI) converted to distances).</p>
</div>
<div class="section level3">
<h3 id="objective">Objective<a class="anchor" aria-label="anchor" href="#objective"></a>
</h3>
<p>Example 1: <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> uses a grid search to
fine-tune hyperparameters for Random Forest and AdaBoost. However, this
process is very time consuming. Additionally, the user is only
restricted to the parameters that are in the grid that
<code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> uses (see supplementary materials in the
<em>aurora</em> paper). However, you can use <code><a href="../reference/get_bags.html">get_bags()</a></code> to
get adjusted datasets and then use your favorite hyperparameter search
method to get the best results.</p>
<p>Example 2: <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> uses four machine learning
algorithms to identify the mislabeled strains. However, these machine
learning algorithms may not be optimal for your dataset and you may want
to explore other options. Machine learning algorithms often learn the
phylogenetic background rather than the existing causal features. To
minimize this issue we will first use <code><a href="../reference/get_bags.html">get_bags()</a></code> to obtain
multiple bootstrapped datasets that are adjusted for phylogenetic
non-independence and then construct corresponding XGBoost models. The
results from these models will be combined to obtain the most probable
causal features and to identify potentially mislabeled strains (a simple
implementation of <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> and
<code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code>).</p>
<p>Example 3: In scientific literature I often encounter plots that are
showing cluster analysis of microbial strains. The results of the
analysis are then displayed in two dimensions. Such plots are supposed
to convince us that <em>i.e.,</em> cluster A is more associated with
trait B. Moreover, some dimensionality reduction techniques, like PCA,
aims to capture the maximum variability in the dataset. However,
population structure and sampling bias can distort such plot. We will
use function <code><a href="../reference/get_bags.html">get_bags()</a></code> to reduce the influence of
population structure and sampling bias. First we will obtain a
population adjusted dataset and then we will calculate Jaccard distance
between all strains. The resulting distance matrix will be plotted to
two dimensions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DalimilBujdos/aurora" class="external-link">aurora</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlr.mlr-org.com" class="external-link">mlr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">randomForest</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dmlc/xgboost" class="external-link">xgboost</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">proxy</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://factominer.free.fr" class="external-link">FactoMineR</a></span><span class="op">)</span></span></code></pre></div>
<p>Now let’s load some data. We will use <em>Limosilactobacillus
reuteri</em> datset. This dataset contains 207 strains from five
different hosts. As a feature table we will use pangenome matrix
produced by <a href="https://github.com/gtonkinhill/panaroo" class="external-link">Panaroo</a>. We will first
need to convert this pangenome into a binary format. Additionally, we
will filter out the most common (present in more than 95% of all
strains) and the rarest (present in less than 10% of all strains)
genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pheno_mat_reuteri"</span><span class="op">)</span> <span class="co"># loads the phenotype for each strain</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bin_mat_reuteri"</span><span class="op">)</span> <span class="co"># loads the pangenome matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"tree_reuteri"</span><span class="op">)</span> <span class="co"># loads the core-genome phylogenetic tree</span></span>
<span></span>
<span><span class="co"># convert the pangenome matrix to binary matrix</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span></span>
<span><span class="va">fut_row_names</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span><span class="op">$</span><span class="va">Gene</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">:</span><span class="op">-</span><span class="fl">14</span><span class="op">]</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/replace.html" class="external-link">replace</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">.</span><span class="op">!=</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">bin_mat</span>, <span class="va">as.numeric</span><span class="op">)</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">fut_row_names</span></span>
<span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">pheno_mat</span><span class="op">$</span><span class="va">ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">no_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span></span>
<span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="va">no_gf</span> <span class="op">*</span> <span class="op">(</span><span class="fl">95</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span> <span class="co"># remove abundat features</span></span>
<span><span class="va">select_ID_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">cutoff</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="va">no_gf</span> <span class="op">*</span> <span class="op">(</span><span class="fl">5</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span> <span class="co"># remove rare features</span></span>
<span><span class="va">non_select_ID_low</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">cutoff</span><span class="op">)</span></span>
<span><span class="va">bin_mat</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span><span class="op">[</span>, <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">select_ID_up</span>, <span class="va">non_select_ID_low</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-1">Example 1<a class="anchor" aria-label="anchor" href="#example-1"></a>
</h2>
<p>With the data loaded we can run <code><a href="../reference/get_bags.html">get_bags()</a></code> to obtain the
bootstrapped dataset. <em>aurora</em> uses two algorithms to bootstrap
the data: <code>phylogenetic_walk</code> and <code>random_walk</code>.
This is controlled by a parameter <code>bagging</code>. If you wish to
find out the distinctions between the two algorithms, please refer to
the supplementary materials of the <em>aurora</em> paper. Both
algorithms are stochastic and therefore you should either repeat the
hyperparameter tuning multiple times (each time with different
bootstrapped dataset) or you can simply set the parameter
<code>bag_size</code> to a large number. This parameter controls how
many strains from each class will be selected to the bootstrapped
dataset. If you want to repeat the parameter search multiple times, you
can get multiple bags with just one run of <code><a href="../reference/get_bags.html">get_bags()</a></code> by
changing the parameter <code>no_rf</code>. In this case, we will set
<code>bag_size</code> to a large number and then use random
hyperparameter search to find the best hyperparameters for random forest
model.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the train dataset</span></span>
<span><span class="va">bag</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_bags.html">get_bags</a></span><span class="op">(</span><span class="va">pheno_mat</span>,</span>
<span>                dist_mat_or_tree <span class="op">=</span> <span class="va">tree</span>, </span>
<span>                bag_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">200</span>, <span class="fl">200</span>, <span class="fl">200</span>, <span class="fl">200</span><span class="op">)</span> <span class="co"># number of strains in each class</span></span>
<span>                <span class="op">)</span></span>
<span><span class="co"># The number of strains in each class should be balanced for Random Forest</span></span>
<span><span class="co"># Get bags also outputs $count which shows the number of strain repetitions</span></span>
<span></span>
<span><span class="co"># Construct new pheno_mat and bin_mat</span></span>
<span><span class="va">boot_pheno_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                             pheno <span class="op">=</span> <span class="va">pheno_mat</span><span class="op">$</span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">pheno_mat</span><span class="op">$</span><span class="va">ids</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">boot_bin_mat</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">boot_bin_mat</span><span class="op">$</span><span class="va">host</span> <span class="op">&lt;-</span> <span class="va">boot_pheno_mat</span><span class="op">$</span><span class="va">pheno</span></span>
<span></span>
<span><span class="co"># gene names from panaroo are not valid column names</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we are ready to run the hyperparameter tuning. Any method can be
used (random search, manual search, gird search, Bayesian optimization,
etc.). In this case, random search will be used. This method is
implemented in package <code>mlr</code>. We need to choose a measurement
of the models’ performance. In real scenario, you might want to use <a href="https://towardsdatascience.com/understanding-auc-roc-curve-68b2303cc9c5" class="external-link">AUC</a>
or an <a href="https://towardsdatascience.com/the-f1-score-bec2bbc38aa6" class="external-link">F1
Score</a> however let’s keep things simple here and use accuracy. Keep
in mind that the grid search that <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> uses takes
a long time to complete (especially if the your dataset has over 500
strains, over 10,000 features and more than two classes). In such a case
you are forced to find your own hyperparameters.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">task</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mlr.mlr-org.com/reference/ClassifTask.html" class="external-link">makeClassifTask</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">boot_bin_mat</span>, target <span class="op">=</span> <span class="st">"host"</span><span class="op">)</span></span>
<span><span class="co"># define hyperparameter search space</span></span>
<span><span class="va">param_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://paramhelpers.mlr-org.com/reference/makeParamSet.html" class="external-link">makeParamSet</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://paramhelpers.mlr-org.com/reference/Param.html" class="external-link">makeIntegerParam</a></span><span class="op">(</span><span class="st">"sampsize"</span>, lower <span class="op">=</span> <span class="fl">50</span>, upper <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://paramhelpers.mlr-org.com/reference/Param.html" class="external-link">makeIntegerParam</a></span><span class="op">(</span><span class="st">"ntree"</span>, lower <span class="op">=</span> <span class="fl">50</span>, upper <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://paramhelpers.mlr-org.com/reference/Param.html" class="external-link">makeIntegerParam</a></span><span class="op">(</span><span class="st">"mtry"</span>, lower <span class="op">=</span> <span class="fl">10</span>, upper <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>, </span>
<span>  <span class="fu"><a href="https://paramhelpers.mlr-org.com/reference/Param.html" class="external-link">makeIntegerParam</a></span><span class="op">(</span><span class="st">"maxnodes"</span>, lower <span class="op">=</span> <span class="fl">4</span>, upper <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mlr.mlr-org.com/reference/makeTuneControlRandom.html" class="external-link">makeTuneControlRandom</a></span><span class="op">(</span>maxit <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># set search method as random</span></span>
<span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mlr.mlr-org.com/reference/tuneParams.html" class="external-link">tuneParams</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://mlr.mlr-org.com/reference/makeLearner.html" class="external-link">makeLearner</a></span><span class="op">(</span><span class="st">"classif.randomForest"</span><span class="op">)</span>,</span>
<span>  <span class="va">task</span>,</span>
<span>  resampling <span class="op">=</span> <span class="fu"><a href="https://mlr.mlr-org.com/reference/makeResampleDesc.html" class="external-link">makeResampleDesc</a></span><span class="op">(</span><span class="st">"CV"</span>, iters <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,  <span class="co"># 4-fold cross-validation</span></span>
<span>  measures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">acc</span><span class="op">)</span>,  <span class="co"># Use accuracy as the performance measure</span></span>
<span>  par.set <span class="op">=</span> <span class="va">param_set</span>,</span>
<span>  control <span class="op">=</span> <span class="va">ctrl</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">best_params</span> <span class="op">&lt;-</span> <span class="va">result</span><span class="op">$</span><span class="va">x</span></span></code></pre></div>
<p>The best hyperparameters are now in list <code>best_params</code>and
we can use it to run <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> (run only with random
forest).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pheno_mat_reuteri"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bin_mat_reuteri"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"tree_reuteri"</span><span class="op">)</span></span>
<span><span class="va">results_aurora</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aurora_pheno.html">aurora_pheno</a></span><span class="op">(</span>pheno_mat <span class="op">=</span> <span class="va">pheno_mat</span>,</span>
<span>                               bin_mat <span class="op">=</span> <span class="va">bin_mat</span>,</span>
<span>                               type_bin_mat <span class="op">=</span> <span class="st">"panaroo"</span>,</span>
<span>                               tree <span class="op">=</span> <span class="va">tree</span>,</span>
<span>                               fit_parameters <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                               sampsize <span class="op">=</span> <span class="va">best_params</span><span class="op">$</span><span class="va">sampsize</span>,</span>
<span>                               mtry <span class="op">=</span> <span class="va">best_params</span><span class="op">$</span><span class="va">mtry</span>,</span>
<span>                               ntree <span class="op">=</span> <span class="va">best_params</span><span class="op">$</span><span class="va">ntree</span>,</span>
<span>                               maxnodes <span class="op">=</span> <span class="va">best_params</span><span class="op">$</span><span class="va">maxnodes</span>,</span>
<span>                               ovr_log_reg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                               adaboost <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                               CART <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                               write_data <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-2">Example 2:<a class="anchor" aria-label="anchor" href="#example-2"></a>
</h2>
<p>In this example we will use the function <code><a href="../reference/get_bags.html">get_bags()</a></code> to
get 10 bootstrapped datasets. This datasets will be used as training
datasets to construct 10 corresponding XGBoost models. We don’t want the
training dataset to be too large. This could lead to extensive strain
repetition which could result in XGBoost models that learned the
features present in those repeating strains.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the train dataset</span></span>
<span><span class="va">bag</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_bags.html">get_bags</a></span><span class="op">(</span><span class="va">pheno_mat</span>,</span>
<span>                dist_mat_or_tree <span class="op">=</span> <span class="va">tree</span>, </span>
<span>                bag_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">20</span>, <span class="fl">20</span>, <span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span>, <span class="co"># number of strains in each class</span></span>
<span>                no_rf <span class="op">=</span> <span class="fl">10</span>, <span class="co"># specify the number of datasets</span></span>
<span>                bagging <span class="op">=</span> <span class="st">"phylogenetic_walk"</span> <span class="co"># specify the bagging algorithm</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p>Now that the 10 training datasets were obtained, we can start
building XGBoost models. Once an XGBoost model is built we can pull out
the importances (feature gains in this case) and we can also use the
model to predict classification probabilities of the original dataset.
In case of a multiclass problem the feature importances will not make
much sense. Thus, let’s reduce the number of classes to two: rodent and
the rest.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make a list that will hold all the feature importances</span></span>
<span><span class="va">importances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># make a list that will hold all the classification probabilities</span></span>
<span><span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># gene names from panaroo are not valid column names</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bin_mat</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">bin_mat_adjust</span> <span class="op">&lt;-</span> <span class="va">bin_mat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_names</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Construct new pheno_mat and bin_mat</span></span>
<span>  <span class="va">boot_pheno_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                             pheno <span class="op">=</span> <span class="va">pheno_mat</span><span class="op">$</span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">pheno_mat</span><span class="op">$</span><span class="va">ids</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">boot_pheno_mat</span><span class="op">$</span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">boot_pheno_mat</span><span class="op">$</span><span class="va">pheno</span> <span class="op">==</span> <span class="st">"rodent"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span> <span class="co"># 1 = rodent, 0 = other host</span></span>
<span>  <span class="va">boot_bin_mat</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">bag</span><span class="op">$</span><span class="va">bags</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># train XGBoost model</span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xgboost/man/xgb.train.html" class="external-link">xgboost</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/proxy/man/dist.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">boot_bin_mat</span><span class="op">)</span>,</span>
<span>                   label <span class="op">=</span> <span class="va">boot_pheno_mat</span><span class="op">$</span><span class="va">pheno</span>,</span>
<span>                   nrounds <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                   objective <span class="op">=</span> <span class="st">"binary:logistic"</span>,</span>
<span>                   verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="va">importances</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/proxy/man/dist.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xgboost/man/xgb.importance.html" class="external-link">xgb.importance</a></span><span class="op">(</span>model <span class="op">=</span> <span class="va">model</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">probs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">bin_mat_adjust</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The last step is to calculate median for all feature importances and
the number of times it was used in the 10 models. Additionally, the
median of the 10 classification probabilities will also be calculated.
This will allow us to identify potentially mislabeled strains.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># analyze the classification probabilities</span></span>
<span><span class="va">medians</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">probs</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">median</span><span class="op">)</span></span>
<span><span class="va">df_probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bin_mat_adjust</span><span class="op">)</span>,</span>
<span>                       probs_med <span class="op">=</span> <span class="va">medians</span>,</span>
<span>                       host <span class="op">=</span> <span class="va">pheno_mat</span><span class="op">$</span><span class="va">pheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># analyze the feature importances</span></span>
<span><span class="va">df_importances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">importances</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_importances</span><span class="op">$</span><span class="va">Feature</span> <span class="op">&lt;-</span> <span class="va">gene_names</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">df_importances</span><span class="op">$</span><span class="va">Feature</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_names</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">df_importances</span><span class="op">$</span><span class="va">Feature</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="va">features</span><span class="op">[</span><span class="va">features</span><span class="op">$</span><span class="va">Freq</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="va">features</span><span class="op">$</span><span class="va">importance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">features</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df_importances</span><span class="op">$</span><span class="va">Gain</span><span class="op">[</span><span class="va">df_importances</span><span class="op">$</span><span class="va">Feature</span> <span class="op">==</span> <span class="va">features</span><span class="op">$</span><span class="va">Var1</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="va">features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">features</span><span class="op">$</span><span class="va">Freq</span>, <span class="va">features</span><span class="op">$</span><span class="va">importance</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="st">"Frequence"</span>, <span class="st">"Importance"</span><span class="op">)</span></span>
<span><span class="co"># print the top 15 features along with their frequences in the 10 models and the median of their importances</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">features</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,<span class="op">]</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Gene
</th>
<th style="text-align:right;">
Frequence
</th>
<th style="text-align:right;">
Importance
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
95
</td>
<td style="text-align:left;">
menC~~~menC_1
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
0.2945214
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
arlR_2~<del>rssB</del><sub>arlR_1</sub>~~regX3
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0.1154578
</td>
</tr>
<tr>
<td style="text-align:left;">
16
</td>
<td style="text-align:left;">
group_1477
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0.0544126
</td>
</tr>
<tr>
<td style="text-align:left;">
23
</td>
<td style="text-align:left;">
group_2082
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0530852
</td>
</tr>
<tr>
<td style="text-align:left;">
102
</td>
<td style="text-align:left;">
ureD1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0394799
</td>
</tr>
<tr>
<td style="text-align:left;">
25
</td>
<td style="text-align:left;">
group_2194
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.1292370
</td>
</tr>
<tr>
<td style="text-align:left;">
22
</td>
<td style="text-align:left;">
group_2077
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0430572
</td>
</tr>
<tr>
<td style="text-align:left;">
35
</td>
<td style="text-align:left;">
group_2524
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0390301
</td>
</tr>
<tr>
<td style="text-align:left;">
101
</td>
<td style="text-align:left;">
rspR
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0114454
</td>
</tr>
<tr>
<td style="text-align:left;">
77
</td>
<td style="text-align:left;">
group_4980
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.1378386
</td>
</tr>
<tr>
<td style="text-align:left;">
84
</td>
<td style="text-align:left;">
group_5689
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0767048
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
azoR
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0223808
</td>
</tr>
<tr>
<td style="text-align:left;">
54
</td>
<td style="text-align:left;">
group_3253
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0221101
</td>
</tr>
<tr>
<td style="text-align:left;">
43
</td>
<td style="text-align:left;">
group_2882
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0205886
</td>
</tr>
<tr>
<td style="text-align:left;">
18
</td>
<td style="text-align:left;">
group_1839
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0201992
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">features</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;                               Gene Frequence Importance</span></span>
<span><span class="co">#&gt; 95                   menC~~~menC_1         8 0.29452140</span></span>
<span><span class="co">#&gt; 2   arlR_2~~~rssB~~~arlR_1~~~regX3         6 0.11545777</span></span>
<span><span class="co">#&gt; 16                      group_1477         6 0.05441258</span></span>
<span><span class="co">#&gt; 23                      group_2082         5 0.05308520</span></span>
<span><span class="co">#&gt; 102                          ureD1         5 0.03947990</span></span>
<span><span class="co">#&gt; 25                      group_2194         4 0.12923699</span></span>
<span><span class="co">#&gt; 22                      group_2077         4 0.04305723</span></span>
<span><span class="co">#&gt; 35                      group_2524         4 0.03903010</span></span>
<span><span class="co">#&gt; 101                           rspR         4 0.01144544</span></span>
<span><span class="co">#&gt; 77                      group_4980         3 0.13783858</span></span>
<span><span class="co">#&gt; 84                      group_5689         3 0.07670478</span></span>
<span><span class="co">#&gt; 4                             azoR         3 0.02238081</span></span>
<span><span class="co">#&gt; 54                      group_3253         3 0.02211014</span></span>
<span><span class="co">#&gt; 43                      group_2882         3 0.02058855</span></span>
<span><span class="co">#&gt; 18                      group_1839         3 0.02019920</span></span></code></pre></div>
<p>These results contain some of the experimentally confirmed rodent
colonization factors and we can thus conclude that the models (at least
partially) were able to identify causal features. In the next cell, we
will plot the classification probabilities. This will allow us to
identify strains that might be mislabeled.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_probs</span> <span class="op">&lt;-</span> <span class="va">df_probs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df_probs</span><span class="op">$</span><span class="va">probs_med</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_probs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_probs</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">probs_med</span>, color <span class="op">=</span> <span class="va">host</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"XGBoost results"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"analysed strains"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Classification probability"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"Host"</span><span class="op">)</span></span></code></pre></div>
<p><img src="get_bags_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Assuming that the model does not overfit or underfit we would
conclude that the majority of <em>L. reuteri</em> rodent strains are
autochthonous in the host while a few strains appear to be mislabeled.
For more thorough discussion on <em>L. reuteri</em> adaptation
mechanisms see the main <em>aurora</em> manuscript.</p>
</div>
<div class="section level2">
<h2 id="example-3">Example 3:<a class="anchor" aria-label="anchor" href="#example-3"></a>
</h2>
<p>In this last example we will compare two plots. One plot will show a
the original <em>L. reuteri</em> dataset, the other will show the
boostrapped <em>L. reuteri</em> dataset. Using <code><a href="../reference/get_bags.html">get_bags()</a></code>
should result in mitigating the bias that originates from population
structure and sampling bias.</p>
<p><img src="get_bags_files/figure-html/figures-side-1.png" width="50%"><img src="get_bags_files/figure-html/figures-side-2.png" width="50%"></p>
<p>The unadjusted plot is not as well separated as the adjusted plot.
Specifically, the porcine strains are intermixed with poultry strains in
the unadjusted plot. As we show in the main <em>aurora</em> paper,
isolates from these two hosts are actually very distinct from each other
but the unadjusted analysis does not communicate this well.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dalimil Bujdoš.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
