<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>panGWAS • aurora</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="panGWAS">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">aurora</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/get_bags.html">get_bags</a></li>
    <li><a class="dropdown-item" href="../articles/outputs.html">outputs</a></li>
    <li><a class="dropdown-item" href="../articles/panGWAS.html">panGWAS</a></li>
    <li><a class="dropdown-item" href="../articles/panGWAS_with_large_dataset.html">panGWAS_with_large_dataset</a></li>
    <li><a class="dropdown-item" href="../articles/setup_conda.html">setup_conda</a></li>
    <li><a class="dropdown-item" href="../articles/SNP_and_kmer_GWAS.html">SNP_and_kmer_GWAS</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DalimilBujdos/aurora/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>panGWAS</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/DalimilBujdos/aurora/blob/HEAD/vignettes/panGWAS.Rmd" class="external-link"><code>vignettes/panGWAS.Rmd</code></a></small>
      <div class="d-none name"><code>panGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Pangenome GWAS (panGWAS) is an analysis method that investigates the
relationship between genes in species’ genofond and observable traits.
PanGWAS relies on the construction of pangenomes and phylogenetic trees
which are the main input into the analysis.</p>
</div>
<div class="section level2">
<h2 id="objective">Objective<a class="anchor" aria-label="anchor" href="#objective"></a>
</h2>
<p>In this vignette you will be introduced to the most common use case
of <em>aurora</em> package. You will try to find out if
<em>Limosilactobacillus reuteri</em> is a host adapted species. To this
end you will use an output from Panaroo (pangenome tool) run with 207
<em>L. reuteri</em> assemblies downloaded from NCBI and a core-genome
phylogenetic tree constructed using IQ-TREE. In the end you will
identify some of the genes that allow <em>L. reuteri</em> to colonize
the rodent gastrointestinal tract.</p>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>First, let’s install and attach all the packages we will need for
this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DalimilBujdos/aurora" class="external-link">aurora</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots" class="external-link">gplots</a></span><span class="op">)</span></span></code></pre></div>
<p>Next let’s load all the data and also create the directory where the
output will be stored.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">save_dir</span> <span class="op">&lt;-</span> <span class="st">"/path/to/your/dir/"</span></span>
<span></span>
<span><span class="co"># this line will not be run because you do not want a vignette to create folders</span></span>
<span><span class="co">#if (!dir.exists(save_dir)) {dir.create(save_dir)}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pheno_mat_reuteri"</span><span class="op">)</span> <span class="co"># loads the phenotype for each strain</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bin_mat_reuteri"</span><span class="op">)</span> <span class="co"># loads the pangenome matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"tree_reuteri"</span><span class="op">)</span> <span class="co"># loads the core-genome phylogenetic tree</span></span></code></pre></div>
<p>You need to make sure that all strains in <code>pheno_mat</code> are
also in the bin_mat and in the tree. Additionally, the strain names
should contain only letters, numbers and special signs “<em>”, ”.”. A
strain name cannot start with ”.” or ”</em>”. Now, the main
<em>aurora</em> function will be executed. This function does not do any
GWAS analysis. The only purpose of it is to find out if the species is
adapted to the supplied phenotype and find which strains are mislabelled
(allochthonous strains). If you don’t want to run it, you can skip this
step and run only <code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code>.</p>
<p>The <em>L. reuteri</em> is a medium-size dataset, and thus there is
no need to change any parameters to optimize <em>aurora’s</em>
performance. Still it might take a few hours for
<code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> to finish. By default
<code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> runs with all machine learning (ML)
algorithms. Python functions are used to run AdaBoost and log
regression. For this reason, one needs to set up a conda environment.
Follow a tutorial in <code>setup_conda</code> to set it up properly and
then, provide an exact system path to the conda environment.
Alternatively, you can set <code>adaboost = FALSE</code> and
<code>ovr_log_reg = FALSE</code> to disable those ML algorithmns. If you
don’t provide any <code>condaenv_path</code> then AdaBoost and log
regression will not be run.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># I want to use all ML algorithmns</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aurora_pheno.html">aurora_pheno</a></span><span class="op">(</span>pheno_mat <span class="op">=</span> <span class="va">pheno_mat</span>,</span>
<span>                        bin_mat <span class="op">=</span> <span class="va">bin_mat</span>,</span>
<span>                        type_bin_mat <span class="op">=</span> <span class="st">"panaroo"</span>,</span>
<span>                        tree <span class="op">=</span> <span class="va">tree</span>,</span>
<span>                        condaenv_path <span class="op">=</span> <span class="st">"/path/to/conda/env"</span>, <span class="co"># an example of a path to conda environment</span></span>
<span>                        save_dir <span class="op">=</span> <span class="va">save_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># I don't want to use AdaBoost and log regression</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aurora_pheno.html">aurora_pheno</a></span><span class="op">(</span>pheno_mat <span class="op">=</span> <span class="va">pheno_mat</span>,</span>
<span>                        bin_mat <span class="op">=</span> <span class="va">bin_mat</span>,</span>
<span>                        type_bin_mat <span class="op">=</span> <span class="st">"panaroo"</span>,</span>
<span>                        tree <span class="op">=</span> <span class="va">tree</span>,</span>
<span>                        save_dir <span class="op">=</span> <span class="va">save_dir</span><span class="op">)</span></span>
<span><span class="co"># lets also save the final list for GWAS analysis later</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">results</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">save_dir</span>, <span class="st">"results_aurora.Rdata"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Since running the function will take a few hours lets load the
precalculated results.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"aurora_pheno_results_reuteri"</span><span class="op">)</span> <span class="co"># these are only truncated results!</span></span></code></pre></div>
<p>When you run <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code> the generated figures are
stored in <code>save_dir</code>. These figures named
“aurora_heatmap_CART_[date].png”,
“aurora_heatmap_median_random_forest_[date].png” and
“aurora_heatmap_sum_random_forest_[date].png” are crucial for
identifying clusters of strains that are not adapted to a phenotype
class. These figures contain heatmaps of distances calculated based on
CART and Random Forest proximities. If the species is adapted to the
phenotype then the clusters should correspond with the observed
phenotype classes. Additionally, the heatmaps contains a column color
strip that show all strains whose predicted class matches the observed.
If the color is lighter then the strain was classified as non-typical.
Blank spots means that the class prediction of the strain was either
INCONCLUSIVE or different than the observed class. The row color strip
shows the observed phenotype class. The color legend of these figures is
stored in
<code>results$results$results_random_forest$legend_heatmap</code> and
<code>results$results$results_CART$legend_heatmap</code>. Next we can
examine the results stored in the list <code>results</code> The list
contains the function call, the supplied <code>bin_mat</code>, the
analysed phenotype classes (phenotypes for short).
<code>results$abundance_filter_removed</code>,
<code>results$acr_filter_removed</code> and
<code>results$hamming_filter_removed</code> contains the number of
features that these filters removed or collapsed.
<code>results$plot_acr_cutoff</code> contains a histogram that shows how
the cutoff for the ancestral reconstruction filter
(<code>ancest_rec_filter</code>).
<code>results$pairwise_dist_outlier</code> contains a histogram that
shows the cutoff applied for <code>reduce_oulier</code>. The fitted
parameters for AdaBoost and Random Forest can be found in
<code>results$fitted_parameters</code>.</p>
<p>The most important objects are stored in
<code>results$results$</code>. The complete list of what this part of
the output contains is in <code>?aurora_pheno()</code>. Let’s explore
the main objects. First, the p-value matrices should be examined.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">func</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># this function will make the results look prettier</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>   <span class="kw">if</span> <span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  <span class="kw">while</span> <span class="op">(</span><span class="va">x</span><span class="op">*</span><span class="va">y</span> <span class="op">&lt;</span> <span class="fl">1</span> <span class="op">||</span> <span class="va">x</span><span class="op">*</span><span class="va">y</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span> <span class="op">*</span> <span class="fl">10</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op">*</span> <span class="va">y</span></span>
<span>  <span class="va">x_rounded</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">x_final</span> <span class="op">&lt;-</span> <span class="va">x_rounded</span> <span class="op">/</span> <span class="va">y</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x_final</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_random_forest</span><span class="op">$</span><span class="va">p_val_mat</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">func</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html" class="external-link">heatmap.2</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>          trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          dendrogram <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          cellnote<span class="op">=</span><span class="va">mat</span>,</span>
<span>          notecol<span class="op">=</span><span class="st">"black"</span>,</span>
<span>          density.info<span class="op">=</span><span class="st">"none"</span>,</span>
<span>          key <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Rowv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Colv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          cexRow <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>          cexCol <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_adaboost</span><span class="op">$</span><span class="va">p_val_mat</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">func</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html" class="external-link">heatmap.2</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>          trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          dendrogram <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          cellnote<span class="op">=</span><span class="va">mat</span>,</span>
<span>          notecol<span class="op">=</span><span class="st">"black"</span>,</span>
<span>          density.info<span class="op">=</span><span class="st">"none"</span>,</span>
<span>          key <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Rowv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Colv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          cexRow <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>          cexCol <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_log_reg</span><span class="op">$</span><span class="va">p_val_mat</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">func</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html" class="external-link">heatmap.2</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>          trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          dendrogram <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          cellnote<span class="op">=</span><span class="va">mat</span>,</span>
<span>          notecol<span class="op">=</span><span class="st">"black"</span>,</span>
<span>          density.info<span class="op">=</span><span class="st">"none"</span>,</span>
<span>          key <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Rowv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Colv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          cexRow <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>          cexCol <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_CART</span><span class="op">$</span><span class="va">p_val_mat</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">func</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html" class="external-link">heatmap.2</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>          trace <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          dendrogram <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          cellnote<span class="op">=</span><span class="va">mat</span>,</span>
<span>          notecol<span class="op">=</span><span class="st">"black"</span>,</span>
<span>          density.info<span class="op">=</span><span class="st">"none"</span>,</span>
<span>          key <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Rowv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          Colv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          cexRow <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>          cexCol <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-5-4.png" width="700"></p>
<p>When two p-values (x -&gt; y and y -&gt; x) are both above 0.05 then
the two phenotype classes should be considered indistinguishable and
vice versa when at least one pairwise p-value is below 0.05 then there
is an adaptation to these two classes. <em>aurora</em> thus correctly
identified <em>L. reuteri</em> as host adapted. Additionally, we should
also confirm that results from all bags are consistent. Let’s plot AUC
values for each tool in each iteration of Outlier calculation phase.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_random_forest</span><span class="op">$</span><span class="va">aucs</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Random Forest AUCs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_adaboost</span><span class="op">$</span><span class="va">aucs</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"AdaBoost AUCs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_log_reg</span><span class="op">$</span><span class="va">aucs</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Log reg AUCs"</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-6-3.png" width="700"></p>
<p>Indeed all of the AUCs are close to 1, therefore, no bag is an
outlier. You might have noticed that results from CART were not plotted.
Individual AUCs are not outputted for each iteration of CART. Instead a
median of all runs for each pair of phenotype classes is calculated.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_CART</span><span class="op">$</span><span class="va">complexity</span><span class="op">$</span><span class="va">AUC</span>, names.arg <span class="op">=</span> <span class="cn">NULL</span>, main <span class="op">=</span> <span class="st">"CART AUCs"</span>, horiz <span class="op">=</span> <span class="cn">T</span>, las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="panGWAS_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>You can see that even a simple CART model does a pretty good job on
this dataset, and all class pairs are close to 1. The fact that a
classifier performs well on a dataset does not mean that there are no
mislabelled strains. You can examine all predicted labels in the
<code>auto_allo_results</code> object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These results are truncated to only the first 10 strains</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_random_forest</span><span class="op">$</span><span class="va">auto_allo_results</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
strain_ids
</th>
<th style="text-align:left;">
poultry
</th>
<th style="text-align:left;">
rodent
</th>
<th style="text-align:left;">
homo sapiens
</th>
<th style="text-align:left;">
pig
</th>
<th style="text-align:left;">
primate
</th>
<th style="text-align:left;">
predicted
</th>
<th style="text-align:left;">
observed
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
homo sapiens
</td>
<td style="text-align:left;">
homo sapiens
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_adaboost</span><span class="op">$</span><span class="va">auto_allo_results</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
strain_ids
</th>
<th style="text-align:left;">
poultry
</th>
<th style="text-align:left;">
rodent
</th>
<th style="text-align:left;">
homo sapiens
</th>
<th style="text-align:left;">
pig
</th>
<th style="text-align:left;">
primate
</th>
<th style="text-align:left;">
predicted
</th>
<th style="text-align:left;">
observed
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
homo sapiens
</td>
<td style="text-align:left;">
homo sapiens
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
INCONCLUSIVE
</td>
<td style="text-align:left;">
INCONCLUSIVE or pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_log_reg</span><span class="op">$</span><span class="va">auto_allo_results</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
strain_ids
</th>
<th style="text-align:left;">
poultry
</th>
<th style="text-align:left;">
rodent
</th>
<th style="text-align:left;">
homo sapiens
</th>
<th style="text-align:left;">
pig
</th>
<th style="text-align:left;">
primate
</th>
<th style="text-align:left;">
predicted
</th>
<th style="text-align:left;">
observed
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry not typical
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent not typical
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent not typical
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent not typical
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
homo sapiens
</td>
<td style="text-align:left;">
homo sapiens
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE not typical
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig not typical
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_CART</span><span class="op">$</span><span class="va">auto_allo_results</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
strain_ids
</th>
<th style="text-align:left;">
poultry
</th>
<th style="text-align:left;">
rodent
</th>
<th style="text-align:left;">
homo sapiens
</th>
<th style="text-align:left;">
pig
</th>
<th style="text-align:left;">
primate
</th>
<th style="text-align:left;">
predicted
</th>
<th style="text-align:left;">
observed
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
GCA_025369755.1_ASM2536975v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent or poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622285.1_ASM2462228v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622305.1_ASM2462230v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_024622315.1_ASM2462231v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_025189695.1_ASM2518969v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
homo sapiens
</td>
<td style="text-align:left;">
homo sapiens
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002159305.1_ASM215930v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_002160565.1_ASM216056v1_genomic
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
poultry
</td>
<td style="text-align:left;">
poultry
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022786355.1_ASM2278635v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022780645.1_ASM2278064v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pig
</td>
<td style="text-align:left;">
pig
</td>
</tr>
<tr>
<td style="text-align:left;">
GCA_022511545.1_ASM2251154v1_genomic
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
rodent
</td>
<td style="text-align:left;">
rodent
</td>
</tr>
</tbody>
</table>
<p>It seems that each algorithm identified only a few strains that are
mislabelled. This is another indication that the species is adapted to
the habitat. However, log regression identified the most mislabelled
strains. If results from one algorithm are different than the rest of
the algorithms, we can choose not to use the results in the subsequent
GWAS analysis. Another reason why you may want to omit some algorithms
is that p-value matrix from a ML algorithm shows that the species is not
adapted and it contradicts results of other ML algorithms. Such a model
may just have a high variance because it was overfitted. Also, if the
AUCs of a ML algorithm are much lower than AUCs of others it might mean
that the model is not complex enough (e.g., it has high bias) and the
algorithm should be omitted.</p>
<p>In the example below we run the GWAS analysis without log regression.
Phylogenetic <code>tree</code>, <code>bin_mat</code> and
<code>pheno_mat</code> needs to be provided again. As we show in the
<em>aurora</em> paper there are some rodent isolates that appear to be
able to colonize primate gut as well. To identify the exclusively
primate colonization factors these strains needs to be removed. All
these <em>L. reuteri</em> strains were labelled as non-typical by Random
Forest. We can thus easily identify them and remove them (argument
<code>remove_strains</code>). Alternatively, you can also choose to
remove all non-typical strains by setting
<code>rm_non_typical = TRUE</code>. <code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code> is a
function with many parameters but one of the most important that the
user should pay attention to is the <code>mode</code>. The mode of
<code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code> can either be “consensus” or “strict”. The
first named always removes less strains than “strict” mode. Consensus
mode removes only strains that were identified as allochthonous by all
ML algorithms used. On the other hand, strict mode removes a strain if
it was identified as mislabelled by at least one ML algorithm.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get all non-typical rodent strains identified by Random Forest</span></span>
<span><span class="va">res_rf</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">results_random_forest</span><span class="op">$</span><span class="va">auto_allo_results</span></span>
<span><span class="va">res_rf_rodent</span> <span class="op">&lt;-</span> <span class="va">res_rf</span><span class="op">[</span><span class="va">res_rf</span><span class="op">$</span><span class="va">observed</span> <span class="op">==</span> <span class="st">"rodent"</span>, <span class="op">]</span></span>
<span><span class="va">ids_rf_not_typ</span> <span class="op">&lt;-</span> <span class="va">res_rf_rodent</span><span class="op">$</span><span class="va">strain_ids</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"not typical"</span>, <span class="va">res_rf_rodent</span><span class="op">$</span><span class="va">rodent</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Run GWAS analysis</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aurora_GWAS.html">aurora_GWAS</a></span><span class="op">(</span>bin_mat <span class="op">=</span> <span class="va">bin_mat</span>,</span>
<span>                   type_bin_mat <span class="op">=</span> <span class="st">"panaroo"</span>,</span>
<span>                   pheno_mat <span class="op">=</span> <span class="va">pheno_mat</span>,</span>
<span>                   tree <span class="op">=</span> <span class="va">tree</span>,</span>
<span>                   remove_strains <span class="op">=</span> <span class="va">ids_rf_not_typ</span>,</span>
<span>                   aurora_results <span class="op">=</span> <span class="va">results</span>,</span>
<span>                   mode <span class="op">=</span> <span class="st">"consensus"</span>,</span>
<span>                   write_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                   use_log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 5 branches were identified as outliers. These branches were reduced </span></span>
<span><span class="co">#&gt; Random forest removes 0 strains </span></span>
<span><span class="co">#&gt; Adaboost removes 1 strains </span></span>
<span><span class="co">#&gt; CART removes 1 strains </span></span>
<span><span class="co">#&gt; Consensus mode removed in total 0 strains. There are 207 strains left. </span></span>
<span><span class="co">#&gt; Finished</span></span></code></pre></div>
<p>The removed strains are in <code>res$removed_strains</code> but the
main results are in <code>res$GWAS_results</code>. This data frame
contains precision, recall, F1 values and standardized residuals for
each gene in each class. The best (most causal) genes will have the
highest F1 value and standardized residual. The colonization factors of
rodent gut are well known. It is the urease operon (ure genes),
serine-rich adhesin (group_2460). Let’s see if these colonization
factors can be found in the GWAS results.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GWAS_rodent</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">GWAS_results</span></span>
<span><span class="co"># Sort the genes based on rodent standardized residuals and give them a rank</span></span>
<span><span class="va">GWAS_rodent</span> <span class="op">&lt;-</span> <span class="va">GWAS_rodent</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">GWAS_rodent</span><span class="op">$</span><span class="va">`rodent std_residual`</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">GWAS_rodent</span><span class="op">$</span><span class="va">rank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rank.html" class="external-link">rank</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">GWAS_rodent</span><span class="op">$</span><span class="va">`rodent std_residual`</span>, ties.method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span></span>
<span><span class="co"># Print the rank of the known colonization factors</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"ure"</span>, <span class="va">GWAS_rodent</span><span class="op">$</span><span class="va">Variant</span><span class="op">)</span> <span class="op">|</span> <span class="va">GWAS_rodent</span><span class="op">$</span><span class="va">Variant</span> <span class="op">==</span> <span class="st">"group_2460"</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">GWAS_rodent</span><span class="op">[</span><span class="va">index</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">29</span><span class="op">)</span><span class="op">]</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Variant
</th>
<th style="text-align:right;">
rank
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ureD1
</td>
<td style="text-align:left;">
ureD1
</td>
<td style="text-align:right;">
3.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureF
</td>
<td style="text-align:left;">
ureF
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureG
</td>
<td style="text-align:left;">
ureG
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureE
</td>
<td style="text-align:left;">
ureE
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureC
</td>
<td style="text-align:left;">
ureC
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureA
</td>
<td style="text-align:left;">
ureA
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureB
</td>
<td style="text-align:left;">
ureB
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
ureI
</td>
<td style="text-align:left;">
ureI
</td>
<td style="text-align:right;">
12.5
</td>
</tr>
<tr>
<td style="text-align:left;">
group_2460
</td>
<td style="text-align:left;">
group_2460
</td>
<td style="text-align:right;">
17.0
</td>
</tr>
</tbody>
</table>
<p>Indeed, we can see that the known colonization factors are among the
top results. However, sorting features based on their standardized
residuals may not be the best approach in all scenarios. Non-causal
genes with low frequency can rank before the causal genes. It is a
common practice in GWAS studies to apply frequency cutoffs.
<em>aurora</em> however uses F1 values which implicitly takes into
account the frequency of the gene. Additionally, the user can choose to
filter the results based on their precision value to remove high
frequency non-causal genes.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GWAS_human</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">GWAS_results</span></span>
<span><span class="va">GWAS_human</span> <span class="op">&lt;-</span> <span class="va">GWAS_human</span><span class="op">[</span><span class="va">GWAS_human</span><span class="op">$</span><span class="va">`homo sapiens precision`</span> <span class="op">&gt;</span> <span class="fl">0.4</span>, <span class="op">]</span></span>
<span><span class="va">GWAS_human</span> <span class="op">&lt;-</span> <span class="va">GWAS_human</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">GWAS_human</span><span class="op">$</span><span class="va">`homo sapiens F1`</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># print the top 10 human colonization factors</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">GWAS_human</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">20</span><span class="op">)</span><span class="op">]</span>, format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"class='table table-striped'"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Variant
</th>
<th style="text-align:right;">
homo sapiens F1
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
xerC_5~<del>xerC_6</del><sub>xerD_2</sub>~~xerC_2
</td>
<td style="text-align:left;">
xerC_5~<del>xerC_6</del><sub>xerD_2</sub>~~xerC_2
</td>
<td style="text-align:right;">
0.8077482
</td>
</tr>
<tr>
<td style="text-align:left;">
group_2589
</td>
<td style="text-align:left;">
group_2589
</td>
<td style="text-align:right;">
0.7485191
</td>
</tr>
<tr>
<td style="text-align:left;">
group_4028
</td>
<td style="text-align:left;">
group_4028
</td>
<td style="text-align:right;">
0.7066599
</td>
</tr>
<tr>
<td style="text-align:left;">
group_4023
</td>
<td style="text-align:left;">
group_4023
</td>
<td style="text-align:right;">
0.7066599
</td>
</tr>
<tr>
<td style="text-align:left;">
group_4017
</td>
<td style="text-align:left;">
group_4017
</td>
<td style="text-align:right;">
0.7066599
</td>
</tr>
<tr>
<td style="text-align:left;">
panP
</td>
<td style="text-align:left;">
panP
</td>
<td style="text-align:right;">
0.6495611
</td>
</tr>
<tr>
<td style="text-align:left;">
group_362
</td>
<td style="text-align:left;">
group_362
</td>
<td style="text-align:right;">
0.6299141
</td>
</tr>
<tr>
<td style="text-align:left;">
group_260
</td>
<td style="text-align:left;">
group_260
</td>
<td style="text-align:right;">
0.6178300
</td>
</tr>
<tr>
<td style="text-align:left;">
tnpR~~~hin
</td>
<td style="text-align:left;">
tnpR~~~hin
</td>
<td style="text-align:right;">
0.6068966
</td>
</tr>
<tr>
<td style="text-align:left;">
group_2912
</td>
<td style="text-align:left;">
group_2912
</td>
<td style="text-align:right;">
0.5843621
</td>
</tr>
</tbody>
</table>
<p>In conclusion, we advise the users run <code><a href="../reference/aurora_pheno.html">aurora_pheno()</a></code>
with all machine learning algorithms and with
<code>phylogenetic_walk</code> if a reliable phylogenetic tree is
available. The user should then experiment with different settings of
<code><a href="../reference/aurora_GWAS.html">aurora_GWAS()</a></code>. I often map the removed strains back to the
phylogenetic tree to see how the allochthonous strains are dispersed
around the tree. You can use <code>res$removed_strains</code> for
this.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dalimil Bujdoš.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
